# Famous Paintings Analysis
This is a **PostgreSQL project** based on [this Kaggle dataset.](https://www.kaggle.com/datasets/mexwell/famous-paintings) We're going to **create a database** from csv files, containing data about famous artists, their works and the museums in which they are exhibited. Then we'll **validate, clean and analyse the dataset to get some insights**.

# Jupyter SQL Setup
First of all, we need to **install and load the extensions** needed **for Jupyter to run SQL** queries.

In [1]:
# Install required packages
!pip install ipython-sql sqlalchemy psycopg2 pandas

# Load SQL extension in Jupyter
%load_ext sql



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


# Connect to PostgreSQL
Then, we **connect to PostgreSQL** (without specifying a database).

In [2]:
from sqlalchemy import create_engine
from getpass import getpass

# Ask for credentials and enter them manually, for privacy reasons
username = input("Enter your PostgreSQL username: ")
password = getpass("Enter your PostgreSQL password: ")

# Connect to PostgreSQL without specifying a database
engine = create_engine(f"postgresql://{username}:{password}@localhost:5432/postgres", isolation_level="AUTOCOMMIT")


Enter your PostgreSQL username: postgres
Enter your PostgreSQL password: ········


# Create and Connect to the new Database
And now we are ready to **create and connect to our new database with** a little help from **SQLAlchemy**.

In [3]:
from sqlalchemy import text

# Create a new database
db_name = "famous_paintings"
with engine.connect() as conn:
    conn.execute(text(f"CREATE DATABASE {db_name};"))

print(f"Database '{db_name}' created successfully.")

# Connect to the new database
engine = create_engine(f"postgresql://{username}:{password}@localhost:5432/{db_name}")


Database 'famous_paintings' created successfully.


# Create Tables and Import Data
As these csv files most probably need cleaning, we're going to import them into tables avoiding any data types or constraints that might prevent rows from being imported. We can alter data types or add constraints later on, if needed. After a quick look at the details of the files on the [Famous Paintings Datacard](https://www.kaggle.com/datasets/mexwell/famous-paintings?select=artist.csv) we can see that **smallint, integer and character varying data types should be used** for our columns. *Note that a **few values in the 'size_id' column of the 'product_size' table force us to use CHARVAR** instead of SMALLINT. Also, **we're not going to use the image_link file as the hyperlinks are invalid**.

In [4]:
# Create database connection
engine = create_engine(f"postgresql://{username}:{password}@localhost:5432/{db_name}")

# Explicitly execute and commit
with engine.connect() as conn:
    conn.execute(text("""
        CREATE TABLE IF NOT EXISTS artist (
            artist_id SMALLINT PRIMARY KEY,
            full_name VARCHAR,
            first_name VARCHAR,
            middle_names VARCHAR, 
            last_name VARCHAR,
            nationality VARCHAR,
            style VARCHAR,
            birth SMALLINT,
            death SMALLINT
        );
        
        CREATE TABLE IF NOT EXISTS canvas_size (
            size_id SMALLINT PRIMARY KEY, 
            width SMALLINT,
            height SMALLINT,
            label VARCHAR
        );
            
        CREATE TABLE IF NOT EXISTS image_link (
            work_id INTEGER,
            url VARCHAR,
            thumbnail_small_url VARCHAR,
            thumbnail_large_url VARCHAR
        );
        
        CREATE TABLE IF NOT EXISTS museum (
            museum_id SMALLINT PRIMARY KEY,
            name VARCHAR,
            address VARCHAR,
            city VARCHAR,
            state VARCHAR,
            postal VARCHAR,
            country VARCHAR,
            phone VARCHAR,
            url VARCHAR
        );
            
        CREATE TABLE IF NOT EXISTS museum_hours (
            museum_id SMALLINT,
            day VARCHAR,
            open VARCHAR,
            close VARCHAR
        );
        
        CREATE TABLE IF NOT EXISTS product_size (
            work_id INTEGER,
            size_id VARCHAR,
            sale_price SMALLINT,
            regular_price SMALLINT
        );
        
        CREATE TABLE IF NOT EXISTS subject (
            work_id INTEGER,
            subject VARCHAR
        );
        
        CREATE TABLE IF NOT EXISTS work (
            work_id INTEGER,
            name VARCHAR,
            artist_id SMALLINT,
            style VARCHAR,
            museum_id SMALLINT
        );
    """))
    conn.commit()  # Ensure changes are saved

print("Tables were successfully created!")


Tables were successfully created!


# Populate tables
Now the tables are ready to host our data. Let's bring them in!

In [5]:
# Define csv files and corresponding table names
csv_table_map = {
    r"C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\artist.csv": "artist",
    r"C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\canvas_size.csv": "canvas_size",
    r"C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\image_link.csv": "image_link",
    r"C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\museum.csv": "museum",
    r"C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\museum_hours.csv": "museum_hours",
    r"C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\product_size.csv": "product_size",
    r"C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\subject.csv": "subject",
    r"C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\work.csv": "work"
}

try:
    with engine.begin() as conn:
        for csv_file, table_name in csv_table_map.items():
            print(f"Copying {csv_file} into {table_name}...")

            with open(csv_file, "r", encoding="utf-8") as f:
                next(f)  # Skip the header row
                conn.connection.cursor().copy_expert(f"COPY {table_name} FROM STDIN WITH CSV HEADER", f)

        print("✅ All CSV files copied successfully!")

except Exception as e:
    print(f"❌ Error: {e}")


Copying C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\artist.csv into artist...
Copying C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\canvas_size.csv into canvas_size...
Copying C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\image_link.csv into image_link...
Copying C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\museum.csv into museum...
Copying C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\museum_hours.csv into museum_hours...
Copying C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\product_size.csv into product_size...
Copying C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\subject.csv into subject...
Copying C:\Users\nikos\Documents\Nikos\Data Science\Datasets\Kaggle\Famous Paintings\work.csv into work...
✅ All CSV files copied successfully!


# Validate Data
We **establish a connection from Jupyter Notebook to the PostgreSQL** database and **then use the magic command %%sql** in the beginning of our cells to **run direct SQL queries**.  
Time to inspect our tables and validate our data!

In [6]:
# Establish connection
connection_string = f"postgresql://{username}:{password}@localhost:5432/{db_name}"
%sql $connection_string

## artist Table
We'll check **unique ids and full names**, strange characters in full names, **inconsistencies** between full names and separate name fields, naming incosistencies in nationalities and styles and **invalid inputs in birth and death** columns. Let's begin!

In [7]:
%%sql

SELECT * FROM artist LIMIT 5;

 * postgresql://postgres:***@localhost:5432/famous_paintings
5 rows affected.


[(501, 'Alexandre Cabanel', 'Alexandre', None, 'Cabanel', 'French', 'Classicist', 1823, 1889),
 (502, 'James Ensor', 'James', None, 'Ensor', 'Belgian', 'Expressionist', 1860, 1949),
 (503, 'Maximilien Luce', 'Maximilien', None, 'Luce', 'French', 'Pointillist', 1858, 1941),
 (504, 'August Macke', 'August', None, 'Macke', 'German', 'Expressionist', 1887, 1914),
 (505, 'Thomas Gainsborough', 'Thomas', None, 'Gainsborough', 'English', 'Rococo', 1727, 1788)]

In [8]:
%%sql

SELECT COUNT(*) AS total_rows, 
       COUNT (DISTINCT artist_id) AS unique_ids, 
       COUNT(DISTINCT full_name) AS unique_full_names,
       COUNT(DISTINCT nationality) AS unique_nationalities,
       COUNT(DISTINCT style) AS unique_styles
FROM artist;


 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[(420, 420, 420, 18, 35)]

The **artist_id and full_name columns only hold unique values** and there seem to be 18 different nationalities and 35 different styles in the artist table. We'll visually **check for naming incosistencies in the nationality and style columns** (the easy part of this validation).
### nationality Column

In [9]:
%%sql

SELECT nationality, COUNT(*)
FROM artist
GROUP BY nationality
ORDER BY nationality;

 * postgresql://postgres:***@localhost:5432/famous_paintings
18 rows affected.


[('American', 83),
 ('Austrian', 6),
 ('Belgian', 4),
 ('Canadian', 3),
 ('Danish', 2),
 ('Dutch', 68),
 ('English', 44),
 ('Flemish', 8),
 ('French', 114),
 ('German', 20),
 ('Irish', 2),
 ('Italian', 26),
 ('Japanese', 2),
 ('Mexican', 1),
 ('Norwegian', 2),
 ('Russian', 14),
 ('Spanish', 10),
 ('Swiss', 11)]

18 different nationalities detected but the **3 most frequent ones (French, American and Dutch) add up to more than half of the total** number of artists. Everything looks fine here. Same procedure for style coming next.
### style Column

In [10]:
%%sql

SELECT style, COUNT(*)
FROM artist
GROUP BY style
ORDER BY style;


 * postgresql://postgres:***@localhost:5432/famous_paintings
35 rows affected.


[('America West', 1),
 ('American West', 3),
 ('Art Nouveau', 5),
 ('Avant-Garde', 5),
 ('Baroque', 54),
 ('Classicist', 14),
 ('Colonial', 11),
 ('Cubist', 8),
 ('Early Renaissance', 2),
 ('Expressionist', 25),
 ('Fauvist', 5),
 ('Folk', 3),
 ('Futurist', 5),
 ('High Renaissance', 5),
 ('Hudson River School', 14),
 ('Impressionist', 41),
 ('Landscape Art', 6),
 ('Mannerist', 3),
 ('Marine Art', 20),
 ('Modern Art', 4),
 ('Nabi', 7),
 ('Naturalist', 6),
 ('Neoclassical', 17),
 ('Northern Renaissance', 6),
 ('Orientalist', 9),
 ('Pointillist', 6),
 ('Portraitist', 18),
 ('Post-Impressionist', 16),
 ('Realist', 35),
 ('Renaissance', 9),
 ('Rococo', 23),
 ('Romantic', 17),
 ('Surrealist', 3),
 ('Symbolist', 12),
 ('Ukiyo-e', 2)]

Obviously **'America West' style is a mistyped 'American West'** so we'll **update** that value.

In [11]:
%%sql

UPDATE artist
SET style = 'American West'
WHERE style = 'America West';

 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[]

### birth, death Columns
We can also **check** whether the **difference between birth and death** year is within a **logical** range, say 18 - 110.

In [12]:
%%sql

SELECT * 
FROM artist
WHERE death - birth < 18 
    OR death - birth > 110;

 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[(596, 'Pieter Bruegel', 'Pieter', None, 'Bruegel', 'Dutch', 'Renaissance', 1525, 1530)]

After a quick search, I found out that **Pieter Bruegel was born sometime between 1525 and 1530** (that's what caused this error) and **died in 1569 so we'll update** the **death** column.

In [13]:
%%sql

UPDATE artist
SET death = 1569
WHERE artist_id = 596;

 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[]

Moving on to the hard part of this validation (that requires of lot of manual work), we'll look at the name fields.
### Name related Columns

In [14]:
%%sql

SELECT artist_id, full_name, first_name, middle_names, last_name
FROM artist
ORDER BY artist_id;


 * postgresql://postgres:***@localhost:5432/famous_paintings
420 rows affected.


[(501, 'Alexandre Cabanel', 'Alexandre', None, 'Cabanel'),
 (502, 'James Ensor', 'James', None, 'Ensor'),
 (503, 'Maximilien Luce', 'Maximilien', None, 'Luce'),
 (504, 'August Macke', 'August', None, 'Macke'),
 (505, 'Thomas Gainsborough', 'Thomas', None, 'Gainsborough'),
 (506, 'Julien Dupré', 'Julien', None, 'Dupre'),
 (507, 'George Catlin', 'George', None, 'Catlin'),
 (508, 'Federico Zandomeneghi', 'Federico', None, 'Zandomeneghi'),
 (509, 'Cornelis Springer', 'Cornelis', None, 'Springer'),
 (510, 'Hans Memling', 'Hans', None, 'Memling'),
 (511, 'Piet Mondrian', 'Piet', None, 'Mondrian'),
 (512, 'Lemuel Francis Abbott', 'Lemuel', 'Francis', 'Abbott'),
 (513, 'Mary Cassatt', 'Mary', None, 'Cassatt'),
 (514, 'Heinrich Campendonk', 'Heinrich', None, 'Campendonk'),
 (515, 'Thomas Cole', 'Thomas', None, 'Cole'),
 (516, 'Thomas Luny', 'Thomas', None, 'Luny'),
 (517, 'Eugène-Louis Boudin', 'Eugene', 'Louis', 'Boudin'),
 (518, 'Hendrick Dubbels', 'Hendrick', None, 'Dubbels'),
 (519, 'Frank Duveneck', 'Frank', None, 'Duveneck'),
 (520, 'Asher Brown Durand', 'Asher', 'Brown', 'Durand'),
 (521, 'George Henry Laporte', 'George', 'Henry', 'Laporte'),
 (522, 'Peter Monamy', 'Peter', None, 'Monamy'),
 (523, 'John Francis Rigaud', 'John', 'Francis', 'Rigaud'),
 (524, 'John William Waterhouse', 'John', 'William', 'Waterhouse'),
 (525, 'Francis William Edmonds', 'Francis', 'William', 'Edmonds'),
 (526, 'Frank Weston Benson', 'Benson', 'Frank', 'Weston'),
 (527, 'Jean Raoux', 'Jean', None, 'Raoux'),
 (528, 'Joseph Ducreux', 'Joseph', None, 'Ducreux'),
 (529, 'Gustaf Wilhelm Palm', 'Gustaf', 'Wilhelm', 'Palm'),
 (530, 'Edward William Cooke', 'Edward', 'William', 'Cooke'),
 (531, 'Camille Pissarro', 'Camille', None, 'Pissarro'),
 (532, 'Pieter Quast', 'Pieter', None, 'Quast'),
 (533, 'George Wesley Bellows', 'George', 'Wesley', 'Bellows'),
 (534, 'Alfred Henry Maurer', 'Alfred', 'Henry', 'Maurer'),
 (535, 'Pierre-Henri De Valenciennes', 'Pierre', 'Henri', 'De Valenciennes'),
 (536, 'Henri Lebasque', 'Henri', None, 'Lebasque'),
 (537, 'Claude Lorrain', 'Claude', None, 'Lorraine'),
 (538, 'Thomas Moran', 'Thomas', None, 'Moran'),
 (539, 'Jean-Baptiste Greuze', 'Jean', 'Baptiste', 'Greuze'),
 (540, 'Fernand Cormon', 'Fernand', None, 'Cormon'),
 (541, 'Pierre Puvis De Chavannes', 'Pierre', 'Puvis', 'De Chavannes'),
 (542, 'Sir Henry Raeburn', 'Henry', None, 'Raeburn'),
 (543, 'Charles Peale Polk', 'Charles', 'Peale', 'Polk'),
 (544, 'Peter Paul Rubens', 'Peter', 'Paul', 'Rubens'),
 (545, 'Odilon Redon', 'Odilon', None, 'Redon'),
 (546, 'Vilhelm Hammershoi', 'Vilhelm', None, 'Hammershoi'),
 (547, 'Édouard Vuillard', 'Edouard', None, 'Vuillard'),
 (548, 'Albrecht Durer', 'Albrecht', None, 'Durer'),
 (549, 'Arnold Böcklin', 'Arnold', None, 'Bocklin'),
 (550, 'Claude Monet', 'Claude', None, 'Monet'),
 (551, 'Sir Lawrence Alma-Tadema', 'Lawrence', 'Alma', 'Tadema'),
 (552, 'Maurice Denis', 'Maurice', None, 'Denis'),
 (553, 'El Greco', 'Dom?nikos', None, 'Theotokópoulos'),
 (554, 'Frans Jansz Post', 'Frans', 'Jansz', 'Post'),
 (555, 'Angel Zárraga', 'Angel', None, 'Zarraga'),
 (556, 'James Tissot', 'James', None, 'Tissot'),
 (557, 'Liubov Popova', 'Liubov', None, 'Popova'),
 (558, 'Sansio Raphael', 'Sansio', None, 'Raphael'),
 (559, 'Albert Marquet', 'Albert', None, 'Marquet'),
 (560, 'Charles Marion Russell', 'Charles', 'Marion', 'Russell'),
 (561, 'Canaletto', 'Giovanni', 'Antionio', 'Canaletto'),
 (562, 'Lyonel Feininger', 'Lyonel', None, 'Feininger'),
 (563, 'Édouard Adam', 'Adam', None, 'Edouard'),
 (564, 'Willem Duyster', 'Willem', None, 'Duyster'),
 (565, 'John James Audubon', 'John', 'James', 'Audubon'),
 (566, 'Charles Towne', 'Charles', None, 'Towne'),
 (567, 'Jean Auguste Ingres', 'Jean', 'Auguste', 'Ingres'),
 (568, 'Theo Van Doesburg', 'Theo', None, 'Van Doesburg'),
 (569, 'Adriaan De Lelie', 'Adriaan', None, 'De Lelie'),
 (570, 'Constantin A. Westchiloff', 'Constantin', 'A', 'Westchiloff'),
 (571, 'Théophile Steinlen', 'Theophile', 'Alexand

It seems that the **full_name column is either the full name (along with Sir or Lord title if any) or the nickname/pseudonym** of the artist, **using the punctuation of his nationality** and the **other name columns** are the **corresponding parts** of the full_name **in English**. However there are **lots of inconsistencies**... I wouldn't be addressing these now, as I don't think they will affect my analysis (I would come back later if needed) but as this is a project for practising my skills I'll give them a go!  
We'll start by **replacing all '-' with ' ' to keep** a **consistent** form.

In [15]:
%%sql

UPDATE artist
SET full_name = REPLACE(full_name, '-', ' '),
    first_name = REPLACE(first_name, '-', ' '),
    middle_names = REPLACE(middle_names, '-', ' '),
    last_name = REPLACE(last_name, '-', ' ')
WHERE full_name LIKE '%-%'  -- this WHERE clause is for optimization reasons (could be omitted)
    OR first_name LIKE '%-%'
    OR middle_names LIKE '%-%'
    OR last_name LIKE '%-%';


 * postgresql://postgres:***@localhost:5432/famous_paintings
34 rows affected.


[]

Next, we'll fix the full_name column and finally the first_name, middle_names and last_name values, based on the full name.
### full_name Column
The following **update, fixes missing middle names, mispelling and lower case** problems in the **full_name** column.  
In case you're wondering, I've done my online search for some of these corrections as I'm certainly not an art expert!

In [16]:
%%sql

UPDATE artist
SET full_name = CASE
                    WHEN artist_id = 570 THEN 'Constantin Alexandrovich Westchiloff'
                    WHEN artist_id = 571 THEN 'Théophile Alexandre Steinlen'
                    WHEN artist_id = 586 THEN 'Gabriël Metsu'
                    WHEN artist_id = 624 THEN 'Theodore Clement Steele'
                    WHEN artist_id = 658 THEN 'George Peter Alexander Healy'
                    WHEN artist_id = 707 THEN 'Laurent De La Hyre'
                    WHEN artist_id = 727 THEN 'Geertgen Tot Sint Jans'
                    WHEN artist_id = 768 THEN 'Caesar Boetius Van Everdingen'
                    WHEN artist_id = 820 THEN 'Edmond Francois Aman Jean'
                    WHEN artist_id = 871 THEN 'Konstantin Alekseyevich Korovin'          
                    WHEN artist_id = 897 THEN 'Élisabeth Vigée Le Brun'
                    WHEN artist_id = 898 THEN 'Joseph Mallord William Turner'
                    ELSE full_name
                END;


 * postgresql://postgres:***@localhost:5432/famous_paintings
420 rows affected.


[]

### first_name Column
We'll be **correcting first names that are actually last names, titles passed as first names, as well as two English versions of first names**.

In [17]:
%%sql

UPDATE artist
SET first_name = CASE
                    WHEN artist_id IN (526, 574, 618, 757, 903) THEN SPLIT_PART(full_name, ' ', 1)
                    WHEN artist_id IN (663, 804, 817) THEN SPLIT_PART(full_name, ' ', 2)
                    WHEN artist_id = 553 THEN 'Domenikos'
                    WHEN artist_id = 563 THEN 'Edouard'
                    ELSE first_name
                END;
                    

 * postgresql://postgres:***@localhost:5432/famous_paintings
420 rows affected.


[]

### middle_name Column
Next we'll **address middle_names problems (lower case, first name or part of last_name, only initials** - when whole name is available, **wrong input**).

In [18]:
%%sql

UPDATE artist
SET middle_names = CASE
                        WHEN artist_id IN (526, 570, 591, 597, 608, 618, 624, 667, 753, 767, 805, 806, 816, 871) THEN SPLIT_PART(full_name, ' ', 2)
                        WHEN artist_id IN (658, 699, 800) THEN SPLIT_PART(full_name, ' ', 2) || ' ' || SPLIT_PART(full_name, ' ', 3)
                        WHEN artist_id IN (663, 703, 727, 729, 780, 783, 795, 797, 804, 817, 832) THEN NULL
                        WHEN artist_id = 561 THEN 'Antonio'
                        ELSE middle_names
                    END; 

 * postgresql://postgres:***@localhost:5432/famous_paintings
420 rows affected.


[]

### last_name Column
**Same procedure for last_name** problems.

In [19]:
%%sql

UPDATE artist
SET last_name = CASE
                    WHEN artist_id IN (537, 563, 757, 770, 903) THEN SPLIT_PART(full_name, ' ', 2)
                    WHEN artist_id IN (526, 618, 663, 667) THEN SPLIT_PART(full_name, ' ', 3)
                    WHEN artist_id IN (631, 703, 727, 729, 742, 783, 795, 797, 832) THEN RIGHT(full_name, LENGTH(full_name) - LENGTH(SPLIT_PART(full_name, ' ', 1)) - 1)
                    WHEN artist_id = 553 THEN 'Theotokopoulos'
                    WHEN artist_id = 561 THEN 'Canal'
                    WHEN artist_id = 705 THEN 'Van Der Kooi'
                    ELSE last_name
                END;

 * postgresql://postgres:***@localhost:5432/famous_paintings
420 rows affected.


[]

### Validate Name related Columns are clean
Let's see what we've accomplished!

In [20]:
%%sql

SELECT artist_id, full_name, first_name, middle_names, last_name 
FROM artist
ORDER BY artist_id;


 * postgresql://postgres:***@localhost:5432/famous_paintings
420 rows affected.


[(501, 'Alexandre Cabanel', 'Alexandre', None, 'Cabanel'),
 (502, 'James Ensor', 'James', None, 'Ensor'),
 (503, 'Maximilien Luce', 'Maximilien', None, 'Luce'),
 (504, 'August Macke', 'August', None, 'Macke'),
 (505, 'Thomas Gainsborough', 'Thomas', None, 'Gainsborough'),
 (506, 'Julien Dupré', 'Julien', None, 'Dupre'),
 (507, 'George Catlin', 'George', None, 'Catlin'),
 (508, 'Federico Zandomeneghi', 'Federico', None, 'Zandomeneghi'),
 (509, 'Cornelis Springer', 'Cornelis', None, 'Springer'),
 (510, 'Hans Memling', 'Hans', None, 'Memling'),
 (511, 'Piet Mondrian', 'Piet', None, 'Mondrian'),
 (512, 'Lemuel Francis Abbott', 'Lemuel', 'Francis', 'Abbott'),
 (513, 'Mary Cassatt', 'Mary', None, 'Cassatt'),
 (514, 'Heinrich Campendonk', 'Heinrich', None, 'Campendonk'),
 (515, 'Thomas Cole', 'Thomas', None, 'Cole'),
 (516, 'Thomas Luny', 'Thomas', None, 'Luny'),
 (517, 'Eugène Louis Boudin', 'Eugene', 'Louis', 'Boudin'),
 (518, 'Hendrick Dubbels', 'Hendrick', None, 'Dubbels'),
 (519, 'Frank Duveneck', 'Frank', None, 'Duveneck'),
 (520, 'Asher Brown Durand', 'Asher', 'Brown', 'Durand'),
 (521, 'George Henry Laporte', 'George', 'Henry', 'Laporte'),
 (522, 'Peter Monamy', 'Peter', None, 'Monamy'),
 (523, 'John Francis Rigaud', 'John', 'Francis', 'Rigaud'),
 (524, 'John William Waterhouse', 'John', 'William', 'Waterhouse'),
 (525, 'Francis William Edmonds', 'Francis', 'William', 'Edmonds'),
 (526, 'Frank Weston Benson', 'Frank', 'Weston', 'Benson'),
 (527, 'Jean Raoux', 'Jean', None, 'Raoux'),
 (528, 'Joseph Ducreux', 'Joseph', None, 'Ducreux'),
 (529, 'Gustaf Wilhelm Palm', 'Gustaf', 'Wilhelm', 'Palm'),
 (530, 'Edward William Cooke', 'Edward', 'William', 'Cooke'),
 (531, 'Camille Pissarro', 'Camille', None, 'Pissarro'),
 (532, 'Pieter Quast', 'Pieter', None, 'Quast'),
 (533, 'George Wesley Bellows', 'George', 'Wesley', 'Bellows'),
 (534, 'Alfred Henry Maurer', 'Alfred', 'Henry', 'Maurer'),
 (535, 'Pierre Henri De Valenciennes', 'Pierre', 'Henri', 'De Valenciennes'),
 (536, 'Henri Lebasque', 'Henri', None, 'Lebasque'),
 (537, 'Claude Lorrain', 'Claude', None, 'Lorrain'),
 (538, 'Thomas Moran', 'Thomas', None, 'Moran'),
 (539, 'Jean Baptiste Greuze', 'Jean', 'Baptiste', 'Greuze'),
 (540, 'Fernand Cormon', 'Fernand', None, 'Cormon'),
 (541, 'Pierre Puvis De Chavannes', 'Pierre', 'Puvis', 'De Chavannes'),
 (542, 'Sir Henry Raeburn', 'Henry', None, 'Raeburn'),
 (543, 'Charles Peale Polk', 'Charles', 'Peale', 'Polk'),
 (544, 'Peter Paul Rubens', 'Peter', 'Paul', 'Rubens'),
 (545, 'Odilon Redon', 'Odilon', None, 'Redon'),
 (546, 'Vilhelm Hammershoi', 'Vilhelm', None, 'Hammershoi'),
 (547, 'Édouard Vuillard', 'Edouard', None, 'Vuillard'),
 (548, 'Albrecht Durer', 'Albrecht', None, 'Durer'),
 (549, 'Arnold Böcklin', 'Arnold', None, 'Bocklin'),
 (550, 'Claude Monet', 'Claude', None, 'Monet'),
 (551, 'Sir Lawrence Alma Tadema', 'Lawrence', 'Alma', 'Tadema'),
 (552, 'Maurice Denis', 'Maurice', None, 'Denis'),
 (553, 'El Greco', 'Domenikos', None, 'Theotokopoulos'),
 (554, 'Frans Jansz Post', 'Frans', 'Jansz', 'Post'),
 (555, 'Angel Zárraga', 'Angel', None, 'Zarraga'),
 (556, 'James Tissot', 'James', None, 'Tissot'),
 (557, 'Liubov Popova', 'Liubov', None, 'Popova'),
 (558, 'Sansio Raphael', 'Sansio', None, 'Raphael'),
 (559, 'Albert Marquet', 'Albert', None, 'Marquet'),
 (560, 'Charles Marion Russell', 'Charles', 'Marion', 'Russell'),
 (561, 'Canaletto', 'Giovanni', 'Antonio', 'Canal'),
 (562, 'Lyonel Feininger', 'Lyonel', None, 'Feininger'),
 (563, 'Édouard Adam', 'Edouard', None, 'Adam'),
 (564, 'Willem Duyster', 'Willem', None, 'Duyster'),
 (565, 'John James Audubon', 'John', 'James', 'Audubon'),
 (566, 'Charles Towne', 'Charles', None, 'Towne'),
 (567, 'Jean Auguste Ingres', 'Jean', 'Auguste', 'Ingres'),
 (568, 'Theo Van Doesburg', 'Theo', None, 'Van Doesburg'),
 (569, 'Adriaan De Lelie', 'Adriaan', None, 'De Lelie'),
 (570, 'Constantin Alexandrovich Westchiloff', 'Constantin', 'Alexandrovich', 'Westchiloff'),
 (571, 'Théophile Alexandre Stein

Everything looks fine now.
## canvas_size Table
Next we're going to **validate values** in the **canvas_size** table.

In [21]:
%%sql

SELECT * 
FROM canvas_size
LIMIT 5;

 * postgresql://postgres:***@localhost:5432/famous_paintings
5 rows affected.


[(24, 24, None, '24" Long Edge'),
 (30, 30, None, '30" Long Edge'),
 (36, 36, None, '36" Long Edge'),
 (40, 40, None, '40" Long Edge'),
 (48, 48, None, '48" Long Edge')]

**NULL values in height** column came up. Let's investigate further.

In [22]:
%%sql

SELECT COUNT(*) AS total_rows, 
       COUNT (DISTINCT size_id) AS unique_ids, 
       COUNT(width) AS non_null_widths,
       COUNT(height) AS non_null_heights,
       COUNT(DISTINCT label) AS unique_labels
FROM canvas_size;


 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[(199, 199, 199, 193, 199)]

**Every row has its own unique id and label**, there are **no NULL values in** the **width** column but there are **6 in** the **height** column. As the labels of the first few rows that include null height values don't give us a clue about their heights, we **can't** really do anything reasonable to **impute them** but at least we can **validate the width's and height's range of values**.

In [23]:
%%sql

SELECT 
    'width' AS column_name, MIN(width), MAX(width)
FROM canvas_size
UNION ALL
SELECT 
    'height', MIN(height), MAX(height) FROM canvas_size;


 * postgresql://postgres:***@localhost:5432/famous_paintings
2 rows affected.


[('width', 15, 96), ('height', 16, 96)]

No suspicious numbers here, so we're moving on.

## museum Table


In [24]:
%%sql

SELECT * FROM museum LIMIT 5;

 * postgresql://postgres:***@localhost:5432/famous_paintings
5 rows affected.


[(31, 'Pushkin State Museum of Fine Arts', '12 Ulitsa Volkhonka', 'Moscow', None, '119019', 'Russia', '+7 495 697-95-78', 'https://pushkinmuseum.art/'),
 (32, 'National Gallery of Victoria', '180 St Kilda Rd', 'Melbourne', 'Victoria', '3004', 'Australia', '+61 (0)3 8620 2222', 'https://www.ngv.vic.gov.au/'),
 (33, 'São Paulo Museum of Art', 'Av. Paulista, 1578 - Bela Vista', 'São Paulo', None, '01310-200', 'Brazil', '+55 11 3149-5959', 'https://masp.org.br/'),
 (34, 'The State Hermitage Museum', 'Palace Square', '2', 'Sankt-Peterburg', '190000', 'Russia', '7 812 710-90-79', 'https://www.hermitagemuseum.org/wps/portal/hermitage/'),
 (35, 'The Metropolitan Museum of Art', '1000 5th Ave', 'New York', 'NY', '10028', 'USA', '+1 212 535-7710', 'https://www.metmuseum.org/')]

**Museum id range is 31 - 86** but we can already see **problems with the address, city and state** columns. Let's see how many museum records we have, in order to decide on how to address these problems.

In [25]:
%%sql

SELECT COUNT(*) AS total_rows, COUNT(DISTINCT museum_id) AS unique_ids
FROM museum;

 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[(56, 56)]

That's an easily manageable amount of rows so we'll **fix errors manually**, one by one.

In [26]:
%%sql

SELECT * FROM museum;

 * postgresql://postgres:***@localhost:5432/famous_paintings
56 rows affected.


[(31, 'Pushkin State Museum of Fine Arts', '12 Ulitsa Volkhonka', 'Moscow', None, '119019', 'Russia', '+7 495 697-95-78', 'https://pushkinmuseum.art/'),
 (32, 'National Gallery of Victoria', '180 St Kilda Rd', 'Melbourne', 'Victoria', '3004', 'Australia', '+61 (0)3 8620 2222', 'https://www.ngv.vic.gov.au/'),
 (33, 'São Paulo Museum of Art', 'Av. Paulista, 1578 - Bela Vista', 'São Paulo', None, '01310-200', 'Brazil', '+55 11 3149-5959', 'https://masp.org.br/'),
 (34, 'The State Hermitage Museum', 'Palace Square', '2', 'Sankt-Peterburg', '190000', 'Russia', '7 812 710-90-79', 'https://www.hermitagemuseum.org/wps/portal/hermitage/'),
 (35, 'The Metropolitan Museum of Art', '1000 5th Ave', 'New York', 'NY', '10028', 'USA', '+1 212 535-7710', 'https://www.metmuseum.org/'),
 (36, 'Museum Folkwang', 'Museumsplatz 1', '45128', 'Essen', None, 'Germany', '49 201 8845000', 'https://www.museum-folkwang.de/en'),
 (37, 'Museum of Grenoble', '5 Pl. de Lavalette', '38000', 'Grenoble', None, 'France', '33 4 76 63 44 44', 'https://www.museedegrenoble.fr/1986-the-museum-in-english.htm'),
 (38, 'Musée des Beaux-Arts de Quimper', '40 Pl. Saint-Corentin', '29000', 'Quimper', None, 'France', '33 2 98 95 45 20', 'https://www.mbaq.fr/en/home-3.html'),
 (39, 'Nelson-Atkins Museum of Art', '4525 Oak St', 'Kansas City', 'MO', '64111', 'USA', '+1 816 751-1278', 'https://www.nelson-atkins.org/'),
 (40, 'Musée du Louvre', 'Rue de Rivoli', '75001', 'Paris', None, 'France', '33 1 40 20 50 50', 'https://www.louvre.fr/en'),
 (41, 'National Maritime Museum', 'Romney Rd', 'London', 'SE10 9NF', None, 'United Kingdom', '44 20 8858 4422', 'https://www.rmg.co.uk/national-maritime-museum'),
 (42, 'Museum of Fine Arts Boston', '465 Huntington Ave', 'Boston', 'MA', '2115', 'USA', '617-267-9300', 'https://www.mfa.org/'),
 (43, 'Rijksmuseum', 'Museumstraat 1', 'Amsterdam', None, '1071 XX', 'Netherlands', '31 20 674 7000', 'https://www.rijksmuseum.nl/en/visit'),
 (44, 'Israel Museum', 'Derech Ruppin 11', 'Jerusalem', None, None, 'Israel', '972 2-670-8811', 'https://www.imj.org.il/en'),
 (45, 'Kunsthaus Zürich', 'Heimplatz 1', 'Zurich', None, '8001', 'Switzerland', '+41 44 253 84 84', 'https://www.kunsthaus.ch/en/'),
 (46, 'National Gallery of Art', 'Constitution Ave. NW', 'Washington', 'DC', '20565', 'USA', '+1 202 737-4215', 'https://www.nga.gov/'),
 (47, 'National Gallery', 'Trafalgar Square', 'London', 'England', 'WC2N 5DN', 'UK', '+44 (0)20 7747 2885', 'https://www.nationalgallery.org.uk/'),
 (48, 'Mauritshuis Museum', 'Plein 29', 'Den Haag', None, '2511', 'Netherlands', '+31 70 302 3456', 'https://www.mauritshuis.nl/en/'),
 (49, "Musée d'Orsay", "1 Rue de la Légion d'Honneur", 'Paris', None, '75007', 'France', '+33 (0)1 40 49 48 14', 'https://www.musee-orsay.fr/en'),
 (50, 'The Prado Museum', 'C. de Ruiz de Alarcón, 23', 'Madrid', None, '28014', 'Spain', '+34 913 30 28 00', 'https://www.museodelprado.es/en'),
 (51, 'The Barnes Foundation', '2025 Benjamin Franklin Pkwy', 'Philadelphia', 'PA', '19130', 'USA', '+1 215 278-7000', 'https://www.barnesfoundation.org/'),
 (52, 'Hungarian National Gallery', 'Szent György tér 2', 'Budapest', None, '1014', 'Hungary', '+36 20 439 7331', 'https://en.mng.hu/'),
 (53, 'Cleveland Museum Of Art', '11150 East Blvd', 'Cleveland', 'OH', '44106', 'USA', '+1 216 421-7350', 'http://clevelandart.org/'),
 (54, 'Museum of Fine Arts, Houston', '1001 Bissonnet St', 'Houston', 'TX', '77005', 'USA', '+1 713 639-7300', 'https://www.mfah.org/'),
 (55, 'The J. Paul Getty Museum', '1200 Getty Center Dr', 'Los Angeles', 'CA', '90049', 'USA', '+1 310 440-7300', 'https://www.getty.edu/museum/'),
 (56, 'Thussen-Bornemisza Museum', 'P.º del Prado, 8', 'Madrid', None, '28014', 'Spain', '+ 34 917 91 13 70', 'https://www.museothyssen.org/en'),
 (57, 'Van Gogh Museum', 'Museumplein 6', 'Amsterdam', None, '1071 DJ', 'Netherlands', '+1 020 570 52 00', 'https://www.vangoghmuseum.nl/en'),
 (58, 'The Phillips Collection', '1600 21st St NW', 'Washington', 'DC', '200

In [27]:
%%sql

WITH updates (museum_id, address, city, state, postal) AS (
    VALUES
        (34, '2 Palace Square', 'Sankt-Peterburg', NULL, '190000'),
        (36, 'Museumsplatz 1', 'Essen', NULL, '45128'),
        (37, '5 Pl. de Lavalette', 'Grenoble', NULL, '38000'),
        (38, '40 Pl. Saint-Corentin', 'Quimper', NULL, '29000'), 
        (40, 'Rue de Rivoli', 'Paris', NULL, '75001'), 
        (41, 'Romney Rd', 'London', NULL, 'SE10 9NF'), 
        (47, 'Trafalgar Square', 'London', NULL, 'WC2N 5DN'), 
        (62, 'Millbank', 'London', NULL, 'SW1P 4RG'),
        (74, 'Houtkampweg 6', 'Otterlo', NULL, '6731 AW'), 
        (75, 'Cathays Park', 'Cardiff', NULL, 'CF10 3NP'),
        (86, 'Somerset House', 'London', 'Strand', 'WC2R 0RN')        
)
UPDATE museum
SET 
    address = updates.address,
    city = updates.city,
    state = updates.state
FROM updates
WHERE museum.museum_id = updates.museum_id;

UPDATE museum
SET country = CASE
                WHEN museum_id IN (41, 47, 62, 78, 86) THEN 'England'
                WHEN museum_id = 75 THEN 'Wales'
                ELSE country
              END;

 * postgresql://postgres:***@localhost:5432/famous_paintings
11 rows affected.
56 rows affected.


[]

In [28]:
%%sql 

SELECT * FROM museum;


 * postgresql://postgres:***@localhost:5432/famous_paintings
56 rows affected.


[(31, 'Pushkin State Museum of Fine Arts', '12 Ulitsa Volkhonka', 'Moscow', None, '119019', 'Russia', '+7 495 697-95-78', 'https://pushkinmuseum.art/'),
 (32, 'National Gallery of Victoria', '180 St Kilda Rd', 'Melbourne', 'Victoria', '3004', 'Australia', '+61 (0)3 8620 2222', 'https://www.ngv.vic.gov.au/'),
 (33, 'São Paulo Museum of Art', 'Av. Paulista, 1578 - Bela Vista', 'São Paulo', None, '01310-200', 'Brazil', '+55 11 3149-5959', 'https://masp.org.br/'),
 (35, 'The Metropolitan Museum of Art', '1000 5th Ave', 'New York', 'NY', '10028', 'USA', '+1 212 535-7710', 'https://www.metmuseum.org/'),
 (39, 'Nelson-Atkins Museum of Art', '4525 Oak St', 'Kansas City', 'MO', '64111', 'USA', '+1 816 751-1278', 'https://www.nelson-atkins.org/'),
 (42, 'Museum of Fine Arts Boston', '465 Huntington Ave', 'Boston', 'MA', '2115', 'USA', '617-267-9300', 'https://www.mfa.org/'),
 (43, 'Rijksmuseum', 'Museumstraat 1', 'Amsterdam', None, '1071 XX', 'Netherlands', '31 20 674 7000', 'https://www.rijksmuseum.nl/en/visit'),
 (44, 'Israel Museum', 'Derech Ruppin 11', 'Jerusalem', None, None, 'Israel', '972 2-670-8811', 'https://www.imj.org.il/en'),
 (45, 'Kunsthaus Zürich', 'Heimplatz 1', 'Zurich', None, '8001', 'Switzerland', '+41 44 253 84 84', 'https://www.kunsthaus.ch/en/'),
 (46, 'National Gallery of Art', 'Constitution Ave. NW', 'Washington', 'DC', '20565', 'USA', '+1 202 737-4215', 'https://www.nga.gov/'),
 (48, 'Mauritshuis Museum', 'Plein 29', 'Den Haag', None, '2511', 'Netherlands', '+31 70 302 3456', 'https://www.mauritshuis.nl/en/'),
 (49, "Musée d'Orsay", "1 Rue de la Légion d'Honneur", 'Paris', None, '75007', 'France', '+33 (0)1 40 49 48 14', 'https://www.musee-orsay.fr/en'),
 (50, 'The Prado Museum', 'C. de Ruiz de Alarcón, 23', 'Madrid', None, '28014', 'Spain', '+34 913 30 28 00', 'https://www.museodelprado.es/en'),
 (51, 'The Barnes Foundation', '2025 Benjamin Franklin Pkwy', 'Philadelphia', 'PA', '19130', 'USA', '+1 215 278-7000', 'https://www.barnesfoundation.org/'),
 (52, 'Hungarian National Gallery', 'Szent György tér 2', 'Budapest', None, '1014', 'Hungary', '+36 20 439 7331', 'https://en.mng.hu/'),
 (53, 'Cleveland Museum Of Art', '11150 East Blvd', 'Cleveland', 'OH', '44106', 'USA', '+1 216 421-7350', 'http://clevelandart.org/'),
 (54, 'Museum of Fine Arts, Houston', '1001 Bissonnet St', 'Houston', 'TX', '77005', 'USA', '+1 713 639-7300', 'https://www.mfah.org/'),
 (55, 'The J. Paul Getty Museum', '1200 Getty Center Dr', 'Los Angeles', 'CA', '90049', 'USA', '+1 310 440-7300', 'https://www.getty.edu/museum/'),
 (56, 'Thussen-Bornemisza Museum', 'P.º del Prado, 8', 'Madrid', None, '28014', 'Spain', '+ 34 917 91 13 70', 'https://www.museothyssen.org/en'),
 (57, 'Van Gogh Museum', 'Museumplein 6', 'Amsterdam', None, '1071 DJ', 'Netherlands', '+1 020 570 52 00', 'https://www.vangoghmuseum.nl/en'),
 (58, 'The Phillips Collection', '1600 21st St NW', 'Washington', 'DC', '20009', 'USA', '+1 202 387-2151', 'https://www.phillipscollection.org/'),
 (59, 'Toledo Museum of Art', '2445 Monroe St', 'Toledo', 'OH', '43620', 'USA', '+1 419 255-8000', 'https://www.toledomuseum.org/'),
 (60, 'Los Angeles County Museum of Art', '5905 Wilshire Blvd', 'Los Angeles', 'CA', '90036', 'USA', '+1 323 857-6000', 'https://www.lacma.org/'),
 (61, 'Solomon R. Guggenheim Museum', '1071 Fifth Avenue', 'New York', 'NY', '10128', 'USA', '+1 212 423-3500', 'https://www.guggenheim.org/'),
 (63, 'Indianapolis Museum of Art', '4000 N Michigan Rd', 'Indianapolis', 'IN', '46208', 'USA', '+1 317923-1331', 'https://discovernewfields.org/'),
 (64, 'Museum of Fine Arts of Nancy', '3 Pl. Stanislas', 'Nancy', None, '54000', 'France', '+33 3 83 85 30 01', 'https://useum.org/museum/Museum-of-Fine-Arts-of-Nancy'),
 (65, 'Fine Arts Museums of San Francisco Legion of Honor', '100 34th  Avenue', 'San Francisco', 'CA', '94121', 'USA', '+1 415 750-3600', 'https://www.famsf.org/'),
 (66, 'Smithsonian American Art Museum', '800 G Street NW', 'Washington', 'DC', '20004', 'USA', '+1

Problems fixed.
## museum_hours Table

In [29]:
%%sql

SELECT * FROM museum_hours LIMIT 5;


 * postgresql://postgres:***@localhost:5432/famous_paintings
5 rows affected.


[(30, 'Monday', '10:30:AM', '05:30:PM'),
 (30, 'Tuesday', '10:30:AM', '05:30:PM'),
 (30, 'Wednesday', '10:30:AM', '05:30:PM'),
 (30, 'Thusday', '10:30:AM', '05:30:PM'),
 (30, 'Friday', '10:30:AM', '05:30:PM')]

We're going to **drop records with museum_id = 30** as it wasn't in our museum_id range, we'll **change** the **datatype of open and close** columns **to TIME in 24h** format to easily calculate museum open hours, **change 'Thusday' to 'Thursday'** and deal with any other inconsistency that will come up! 

In [30]:
%%sql
DELETE FROM museum_hours
WHERE museum_id = 30;

ALTER TABLE museum_hours 
ALTER COLUMN open TYPE TIME USING TO_TIMESTAMP(open, 'HH:MI:PM')::TIME;
        
ALTER TABLE museum_hours 
ALTER COLUMN close TYPE TIME USING TO_TIMESTAMP(close, 'HH:MI:PM')::TIME;

 * postgresql://postgres:***@localhost:5432/famous_paintings
6 rows affected.
Done.
Done.


[]

In [31]:
%%sql

UPDATE museum_hours
SET day = 'Thursday'
WHERE day = 'Thusday';

 * postgresql://postgres:***@localhost:5432/famous_paintings
16 rows affected.


[]

In [32]:
%%sql

SELECT * FROM museum_hours
ORDER BY museum_id
LIMIT 20;


 * postgresql://postgres:***@localhost:5432/famous_paintings
20 rows affected.


[(31, 'Sunday', datetime.time(11, 0), datetime.time(18, 0)),
 (31, 'Friday', datetime.time(11, 0), datetime.time(18, 0)),
 (31, 'Tuesday', datetime.time(11, 0), datetime.time(18, 0)),
 (31, 'Saturday', datetime.time(11, 0), datetime.time(18, 0)),
 (31, 'Wednesday', datetime.time(11, 0), datetime.time(18, 0)),
 (31, 'Thursday', datetime.time(11, 0), datetime.time(18, 0)),
 (31, 'Monday', datetime.time(11, 0), datetime.time(18, 0)),
 (32, 'Sunday', datetime.time(10, 0), datetime.time(17, 0)),
 (32, 'Saturday', datetime.time(10, 0), datetime.time(17, 0)),
 (32, 'Friday', datetime.time(10, 0), datetime.time(17, 0)),
 (32, 'Wednesday', datetime.time(10, 0), datetime.time(17, 0)),
 (32, 'Thursday', datetime.time(10, 0), datetime.time(17, 0)),
 (32, 'Tuesday', datetime.time(10, 0), datetime.time(17, 0)),
 (32, 'Monday', datetime.time(10, 0), datetime.time(17, 0)),
 (33, 'Wednesday', datetime.time(10, 0), datetime.time(18, 0)),
 (33, 'Tuesday', datetime.time(10, 0), datetime.time(20, 0)),
 (33, 'Sunday', datetime.time(10, 0), datetime.time(18, 0)),
 (33, 'Saturday', datetime.time(10, 0), datetime.time(18, 0)),
 (33, 'Friday', datetime.time(10, 0), datetime.time(18, 0)),
 (33, 'Thursday', datetime.time(10, 0), datetime.time(18, 0))]

Looks ok.
## product_size Table

In [33]:
%%sql

SELECT * FROM product_size LIMIT 5;

 * postgresql://postgres:***@localhost:5432/famous_paintings
5 rows affected.


[(160228, '30', 95, 95),
 (160236, '24', 85, 85),
 (160236, '30', 95, 95),
 (160244, '24', 85, 85),
 (160244, '30', 95, 95)]

In [34]:
%%sql

SELECT COUNT(*) AS total, COUNT(size_id) AS size_col, COUNT(sale_price) AS sale_col, COUNT(regular_price) AS regular_col, MIN(sale_price) AS min_sale, MAX(sale_price) AS max_sale, MIN(regular_price) AS min_regular, MAX(regular_price) AS max_regular FROM product_size;

 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[(110346, 110346, 110346, 110346, 10, 1115, 75, 2235)]

**No NULL values**, but let's **check** if there are **unique work_id - size_id pairs**.

In [35]:
%%sql

SELECT work_id, size_id, COUNT(*) AS duplicate_count
FROM product_size
GROUP BY work_id, size_id
HAVING COUNT(*) > 1;

 * postgresql://postgres:***@localhost:5432/famous_paintings
685 rows affected.


[(125785, '4030', 2),
 (135881, '3648', 2),
 (31780, '36', 2),
 (124542, '5436', 2),
 (181752, '3040', 2),
 (181540, '3046', 2),
 (181547, '2436', 2),
 (181412, '4848', 2),
 (124533, '7248', 2),
 (147753, '2430', 2),
 (122171, '2632', 2),
 (124470, '4630', 2),
 (194092, '2430', 2),
 (181483, '3030', 2),
 (181761, '2632', 2),
 (124488, '6040', 2),
 (181909, '3040', 2),
 (181891, '2430', 2),
 (194169, '30', 2),
 (125774, '6048', 2),
 (147806, '3654', 2),
 (194162, '6048', 2),
 (181363, '3040', 2),
 (147753, '4860', 2),
 (181805, '2436', 2),
 (122204, '5468', 2),
 (122171, '2936', 2),
 (147775, '3040', 2),
 (147786, '3040', 2),
 (135936, '2632', 2),
 (31778, '2936', 2),
 (181540, '4872', 2),
 (124470, '30', 2),
 (122224, '3046', 2),
 (181526, '4872', 2),
 (181761, '4860', 2),
 (122182, '3040', 2),
 (124578, '6040', 2),
 (147797, '3046', 2),
 (122204, '2632', 2),
 (181882, '2430', 2),
 (125796, '3629', 2),
 (147764, '1620', 2),
 (194176, '36', 2),
 (194155, '36', 2),
 (135892, '4860', 2),
 (147786, '4860', 2),
 (122204, '3648', 2),
 (122193, '1620', 2),
 (181340, '3040', 2),
 (122691, '5468', 2),
 (135969, '2632', 2),
 (135914, '3648', 2),
 (181441, '4060', 2),
 (124569, '4630', 2),
 (147806, '3046', 2),
 (122691, '2430', 2),
 (181798, '4060', 2),
 (135980, '3648', 2),
 (194101, '36', 2),
 (181457, '3040', 2),
 (135903, '2936', 2),
 (124533, '6040', 2),
 (125785, '30', 2),
 (125752, '4836', 2),
 (181569, '3046', 2),
 (194083, '2632', 2),
 (147786, '2936', 2),
 (181873, '3648', 2),
 (124587, '7248', 2),
 (23448, '4030', 2),
 (135914, '2430', 2),
 (194130, '5468', 2),
 (194121, '30', 2),
 (135969, '3648', 2),
 (181848, '3654', 2),
 (194121, '3648', 2),
 (124515, '5436', 2),
 (122691, '2024', 2),
 (124515, '4630', 2),
 (122193, '3040', 2),
 (124533, '4630', 2),
 (125763, '3024', 2),
 (23448, '20', 2),
 (181499, '2430', 2),
 (124506, '6040', 2),
 (124515, '6040', 2),
 (122204, '2024', 2),
 (181491, '6060', 2),
 (194083, '4860', 2),
 (147742, '2430', 2),
 (181932, '4860', 2),
 (147764, '2430', 2),
 (124470, '3624', 2),
 (125807, '6048', 2),
 (125752, '3024', 2),
 (124497, '7248', 2),
 (124470, '7248', 2),
 (122182, '5468', 2),
 (124551, '6040', 2),
 (24536, '1620', 2),
 (124587, '3624', 2),
 (181752, '5468', 2),
 (181882, '4860', 2),
 (194169, '36', 2),
 (124551, '7248', 2),
 (125752, '30', 2),
 (122691, '3648', 2),
 (124524, '7248', 2),
 (124488, '3624', 2),
 (181891, '4860', 2),
 (194074, '2936', 2),
 (181533, '3046', 2),
 (181547, '3046', 2),
 (181441, '3654', 2),
 (181594, '3040', 2),
 (181925, '3654', 2),
 (181848, '3046', 2),
 (122204, '3040', 2),
 (125796, '6048', 2),
 (122662, '3024', 2),
 (24532, '4630', 2),
 (181508, '3648', 2),
 (125796, '4836', 2),
 (181499, '3648', 2),
 (194101, '3040', 2),
 (125785, '3226', 2),
 (124533, '5436', 2),
 (194092, '3040', 2),
 (147731, '3040', 2),
 (125763, '4836', 2),
 (181594, '4860', 2),
 (124470, '20', 2),
 (194162, '30', 2),
 (181805, '4872', 2),
 (31778, '3040', 2),
 (181363, '5468', 2),
 (181526, '3654', 2),
 (147815, '6854', 2),
 (181396, '4060', 2),
 (181390, '4872', 2),
 (194130, '2936', 2),
 (194121, '5468', 2),
 (181533, '4060', 2),
 (194169, '4030', 2),
 (124542, '3624', 2),
 (147731, '1620', 2),
 (122662, '2016', 2),
 (124497, '6040', 2),
 (194130, '4860', 2),
 (194169, '4836', 2),
 (125774, '24', 2),
 (194148, '36', 2),
 (135925, '3648', 2),
 (194110, '2936', 2),
 (147797, '2030', 2),
 (181932, '2936', 2),
 (122171, '2430', 2),
 (135958, '2936', 2),
 (194074, '36', 2),
 (122691, '2936', 2),
 (181499, '3040', 2),
 (181318, '4860', 2),
 (122182, '1620', 2),
 (125807, '3226', 2),
 (23448, '3629', 2),
 (125785, '3629', 2),
 (135980, '2632', 2),
 (194083, '5468', 2),
 (147764, '3648', 2),
 (122204, '2430', 2),
 (181873, '2936', 2),
 (147815, '3629', 2),
 (181925, '3046', 2),
 (147775, '2430', 2),
 (194121, '4860', 2),
 (135947, '5468', 2),
 (181882, '2936', 2),
 (181918, '4060', 2),
 (122224, '2436', 2),
 (194074, '30', 2),
 (181562, '3046', 2),
 (181412, '6060', 2),
 (122224, '4060'

**Lots of duplicates** so we must get rid of them. To achieve this, we'll **create a new table (product_size_clean) keeping only** rows with **unique work_id - size_id pairs** and **then rename it to product_size after dropping** the **original product_size table*.

In [36]:
%%sql

CREATE TABLE product_size_clean AS 
SELECT DISTINCT ON (work_id, size_id) *
FROM product_size;

DROP TABLE product_size;

ALTER TABLE product_size_clean RENAME TO product_size;

 * postgresql://postgres:***@localhost:5432/famous_paintings
109660 rows affected.
Done.
Done.


[]

## subject Table

In [37]:
%%sql

SELECT * FROM subject LIMIT 5;

 * postgresql://postgres:***@localhost:5432/famous_paintings
5 rows affected.


[(160236, 'Still-Life'),
 (160244, 'Still-Life'),
 (160252, 'Still-Life'),
 (160260, 'Still-Life'),
 (160268, 'Still-Life')]

In [38]:
%%sql

SELECT COUNT(*) AS total_rows, 
       COUNT(work_id) AS non_null_work_ids, 
       COUNT (DISTINCT work_id) AS unique_work_ids,
       COUNT(subject) AS subjects,
       COUNT (DISTINCT subject) AS unique_subjects
FROM subject;

 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[(6770, 6770, 6007, 6770, 29)]

There are **many duplicates in the work_id** column which must be eliminated. Let's first take a look at them.

In [39]:
%%sql

SELECT s.*
FROM subject s
JOIN (
    SELECT work_id
    FROM subject
    GROUP BY work_id
    HAVING COUNT(*) > 1
) duplicates ON s.work_id = duplicates.work_id
ORDER BY work_id;

 * postgresql://postgres:***@localhost:5432/famous_paintings
1497 rows affected.


[(389, 'Flowers'),
 (389, 'Gardens'),
 (414, 'Lovers'),
 (414, 'Gardens'),
 (466, 'Spring'),
 (466, 'Bridges'),
 (469, 'Flowers'),
 (469, 'Gardens'),
 (470, 'Flowers'),
 (470, 'Gardens'),
 (471, 'Gardens'),
 (471, 'Spring'),
 (471, 'Flowers'),
 (479, 'Gardens'),
 (479, 'Flowers'),
 (523, 'Gardens'),
 (523, 'Flowers'),
 (544, 'Gardens'),
 (544, 'Flowers'),
 (762, 'Still-Life'),
 (762, 'Flowers'),
 (768, 'Flowers'),
 (768, 'Still-Life'),
 (770, 'Flowers'),
 (770, 'Still-Life'),
 (780, 'Architectures'),
 (780, 'Churches/Temples/Mosques'),
 (787, 'Spring'),
 (787, 'Flowers'),
 (907, 'Gardens'),
 (907, 'Flowers'),
 (950, 'Autumn/Fall'),
 (950, 'Summer'),
 (1368, 'Jesus Christ'),
 (1368, 'Christianity'),
 (1568, 'Nude'),
 (1568, 'Still-Life'),
 (1580, 'Bridges'),
 (1580, 'Rivers/Lakes'),
 (1940, 'Jesus Christ'),
 (1940, 'Christianity'),
 (1960, 'Christianity'),
 (1960, 'Jesus Christ'),
 (1962, 'Jesus Christ'),
 (1962, 'Christianity'),
 (2024, 'Gardens'),
 (2024, 'Autumn/Fall'),
 (2028, 'Winter'),
 (2028, 'Rivers/Lakes'),
 (2034, 'Seascapes'),
 (2034, 'Autumn/Fall'),
 (2038, 'Rivers/Lakes'),
 (2038, 'Water Lilies'),
 (2038, 'Autumn/Fall'),
 (2062, 'Rivers/Lakes'),
 (2062, 'Bridges'),
 (2080, 'Rivers/Lakes'),
 (2080, 'Bridges'),
 (2084, 'Bridges'),
 (2084, 'Rivers/Lakes'),
 (2088, 'Bridges'),
 (2088, 'Rivers/Lakes'),
 (2090, 'Bridges'),
 (2090, 'Rivers/Lakes'),
 (2098, 'Spring'),
 (2098, 'Seascapes'),
 (2114, 'Autumn/Fall'),
 (2114, 'Flowers'),
 (2130, 'Gardens'),
 (2130, 'Flowers'),
 (2172, 'Architectures'),
 (2172, 'Rivers/Lakes'),
 (2206, 'Gardens'),
 (2206, 'Spring'),
 (2316, 'Water Lilies'),
 (2316, 'Rivers/Lakes'),
 (2354, 'Winter'),
 (2354, 'Rivers/Lakes'),
 (2356, 'Rivers/Lakes'),
 (2356, 'Spring'),
 (2364, 'Still-Life'),
 (2364, 'Flowers'),
 (2376, 'Rivers/Lakes'),
 (2376, 'Winter'),
 (2386, 'Gardens'),
 (2386, 'Seascapes'),
 (2412, 'Rivers/Lakes'),
 (2412, 'Bridges'),
 (2418, 'Water Lilies'),
 (2418, 'Bridges'),
 (2418, 'Rivers/Lakes'),
 (2442, 'Flowers'),
 (2442, 'Gardens'),
 (2444, 'Bridges'),
 (2444, 'Water Lilies'),
 (2460, 'Gardens'),
 (2460, 'Flowers'),
 (2462, 'Gardens'),
 (2462, 'Flowers'),
 (2476, 'Rivers/Lakes'),
 (2476, 'Bridges'),
 (2476, 'Water Lilies'),
 (2506, 'Flowers'),
 (2506, 'Gardens'),
 (2518, 'Bridges'),
 (2518, 'Rivers/Lakes'),
 (2536, 'Water Lilies'),
 (2536, 'Rivers/Lakes'),
 (2538, 'Rivers/Lakes'),
 (2538, 'Water Lilies'),
 (2540, 'Water Lilies'),
 (2540, 'Rivers/Lakes'),
 (2540, 'Bridges'),
 (2646, 'Rivers/Lakes'),
 (2646, 'Spring'),
 (2646, 'Water Lilies'),
 (2650, 'Water Lilies'),
 (2650, 'Rivers/Lakes'),
 (2652, 'Bridges'),
 (2652, 'Water Lilies'),
 (2652, 'Rivers/Lakes'),
 (2654, 'Water Lilies'),
 (2654, 'Rivers/Lakes'),
 (2656, 'Water Lilies'),
 (2656, 'Rivers/Lakes'),
 (2658, 'Rivers/Lakes'),
 (2658, 'Water Lilies'),
 (2754, 'Musics'),
 (2754, 'Cafes/Bars'),
 (2876, 'Musics'),
 (2876, 'Cafes/Bars'),
 (2986, 'Abstract/Modern Art'),
 (2986, 'Churches/Temples/Mosques'),
 (3018, 'Abstract/Modern Art'),
 (3018, 'Gardens'),
 (3064, 'Abstract/Modern Art'),
 (3064, 'Churches/Temples/Mosques'),
 (3080, 'Churches/Temples/Mosques'),
 (3080, 'Abstract/Modern Art'),
 (3096, 'Abstract/Modern Art'),
 (3096, 'Portraits'),
 (3098, 'Abstract/Modern Art'),
 (3098, 'Portraits'),
 (3104, 'Abstract/Modern Art'),
 (3104, 'Churches/Temples/Mosques'),
 (3114, 'Abstract/Modern Art'),
 (3114, 'Horses'),
 (3130, 'Abstract/Modern Art'),
 (3130, 'Churches/Temples/Mosques'),
 (3134, 'Winter'),
 (3134, 'Abstract/Modern Art'),
 (3187, 'Jesus Christ'),
 (3187, 'Christianity'),
 (3189, 'Christianity'),
 (3189, 'Jesus Christ'),
 (3189, 'Seascapes'),
 (3191, 'Christianity'),
 (3191, 'Jesus Christ'),
 (3333, 'Jesus Christ'),
 (3333, 'Christianity'),
 (3421, 'Gardens'),
 (3421, 'Flowers'),
 (3449, 'Gardens'),
 (3449, 'Flowers'),
 (3875, 'Cafes/Bars'),
 (3875, 'Dancers'),
 (3875, 'Lovers'),
 (4091, 'Nude'),
 (4091, 'Flowers'),
 (4107, 'Gardens'),
 (4107, 'Flowers'),
 (4113, 'Nude'),
 (4113, 'Portraits'),
 (4179, 'Flowers'),
 (4179,

It seems that the **same theme** has been **classified into different categories** for different products, so **we'll arbitrarily keep one for each work_id**. To achieve this, we'll **create a new table keeping only rows with unique work_id** and **then rename it to subject after dropping the original subject table**.

In [40]:
%%sql

CREATE TABLE subject_clean AS 
SELECT DISTINCT ON (work_id) * 
FROM subject;

DROP TABLE subject;

ALTER TABLE subject_clean RENAME TO subject;

 * postgresql://postgres:***@localhost:5432/famous_paintings
6007 rows affected.
Done.
Done.


[]

We'll also look at the subject categories to **check for mispellings**.

In [41]:
%%sql

SELECT COUNT(*), subject
FROM subject
GROUP BY subject
ORDER BY subject;

 * postgresql://postgres:***@localhost:5432/famous_paintings
29 rows affected.


[(371, 'Abstract/Modern Art'),
 (117, 'Architectures'),
 (67, 'Autumn/Fall'),
 (106, 'Bridges'),
 (47, 'Cafes/Bars'),
 (164, 'Christianity'),
 (61, 'Churches/Temples/Mosques'),
 (97, 'Dancers'),
 (25, 'Deers'),
 (93, 'Dogs'),
 (414, 'Flowers'),
 (234, 'Gardens'),
 (261, 'Horses'),
 (62, 'Jesus Christ'),
 (436, 'Landscape Art'),
 (55, 'Lovers'),
 (234, 'Marine Art/Maritime'),
 (130, 'Musics'),
 (512, 'Nude'),
 (1021, 'Portraits'),
 (393, 'Rivers/Lakes'),
 (286, 'Seascapes'),
 (106, 'Spring'),
 (344, 'Still-Life'),
 (90, 'Summer'),
 (29, 'Tigers'),
 (45, 'U.S. Presidents'),
 (46, 'Water Lilies'),
 (161, 'Winter')]

**No problems** detected.
## work Table

In [42]:
%%sql

SELECT * FROM work LIMIT 5;

 * postgresql://postgres:***@localhost:5432/famous_paintings
5 rows affected.


[(160236, "Still Life with Fruit and a Beaker on a Cock's Foot", 615, 'Baroque', 43),
 (160244, 'Still Life with Fruit and a Goldfinch', 615, 'Baroque', 43),
 (160252, 'Still Life with Fruit and Oysters', 615, 'Baroque', 43),
 (160260, 'Still Life with Fruit, Oysters, and a Porcelain Bowl', 615, 'Baroque', 43),
 (160268, 'The Overturned Bouquet', 615, 'Baroque', 43)]

In [43]:
%%sql

SELECT COUNT(*) AS total_rows, 
       COUNT(work_id) AS non_null_work_ids,
       COUNT (DISTINCT work_id) AS unique_work_ids,
       COUNT(name) AS names,
       COUNT(artist_id) AS artists,
       COUNT(DISTINCT artist_id) AS unique_artists,
       COUNT(style) AS styles, 
       COUNT(museum_id) AS museum_ids,
       COUNT (DISTINCT museum_id) AS unique_museum_ids
FROM work;

 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[(14775, 14775, 14715, 14775, 14775, 421, 13489, 4552, 57)]

We notice that **many works lack style** value and **most aren't exhibited in museums** (most probably) or this info is missing. Also, as there are **421 unique_artists and 57 unique_ids instead of 420 and 56** I guess that museum_id = 30 is again present in this table and should be deleted too as well as the odd artist_id. But the most important observation is that **quite a few work_id values aren't unique** which means we have duplicate values in the work_id column. Let's see those records.

In [44]:
%%sql

SELECT w.*
FROM work w
JOIN (
    SELECT work_id
    FROM work
    GROUP BY work_id
    HAVING COUNT(*) > 1
) duplicates ON w.work_id = duplicates.work_id
ORDER BY work_id;


 * postgresql://postgres:***@localhost:5432/famous_paintings
119 rows affected.


[(122662, 'The Dunes at Camiers', 862, 'Realism', None),
 (122662, 'The Dunes at Camiers', 862, 'Realism', None),
 (122691, 'Landscape with a Sunlit Stream', 862, 'Realism', None),
 (122691, 'Landscape with a Sunlit Stream', 862, 'Realism', None),
 (181318, 'Maroon, Pink and Shade of Red', 770, None, None),
 (181318, 'Maroon, Pink and Shade of Red', 770, None, None),
 (181331, 'Blue, Orange and Gold', 770, None, None),
 (181331, 'Blue, Orange and Gold', 770, None, None),
 (181340, 'Untitled I', 770, None, None),
 (181340, 'Untitled I', 770, None, None),
 (181349, 'Green and Maroon', 770, None, None),
 (181349, 'Green and Maroon', 770, None, None),
 (181356, 'Blue, Dark Blue and Orange', 770, None, None),
 (181356, 'Blue, Dark Blue and Orange', 770, None, None),
 (181363, 'Yellow and Pink', 770, None, None),
 (181363, 'Yellow and Pink', 770, None, None),
 (181390, 'Orange, Yellow and Red', 770, None, None),
 (181390, 'Orange, Yellow and Red', 770, None, None),
 (181396, 'Pink Stripe on Light Yellow', 770, None, None),
 (181396, 'Pink Stripe on Light Yellow', 770, None, None),
 (181403, 'Black Stripe, Red Stripe', 770, None, None),
 (181403, 'Black Stripe, Red Stripe', 770, None, None),
 (181412, 'Orange, Black and Maroon', 770, None, None),
 (181412, 'Orange, Black and Maroon', 770, None, None),
 (181420, 'Bright Yellow and Teal Blue', 770, None, None),
 (181420, 'Bright Yellow and Teal Blue', 770, None, None),
 (181427, 'Gold, Red Pink and Black', 770, None, None),
 (181427, 'Gold, Red Pink and Black', 770, None, None),
 (181434, 'Gray, Black and Orange', 770, None, None),
 (181434, 'Gray, Black and Orange', 770, None, None),
 (181441, 'Pastel Blue and Red Stripe', 770, None, None),
 (181441, 'Pastel Blue and Red Stripe', 770, None, None),
 (181448, 'Emerald Green and Blue Stripe', 770, None, None),
 (181448, 'Emerald Green and Blue Stripe', 770, None, None),
 (181457, 'Pastel Shades', 770, None, None),
 (181457, 'Pastel Shades', 770, None, None),
 (181475, 'Brown, Rust and Moss Green', 770, None, None),
 (181475, 'Brown, Rust and Moss Green', 770, None, None),
 (181483, 'Purple, Dark Brown and Rust', 770, None, None),
 (181483, 'Purple, Dark Brown and Rust', 770, None, None),
 (181491, 'Untitled II', 770, None, None),
 (181491, 'Untitled II', 770, None, None),
 (181499, 'Maroon Red, Pink on Orange Yellow', 770, None, None),
 (181499, 'Maroon Red, Pink on Orange Yellow', 770, None, None),
 (181508, 'Pastel Pinks on Dark Gray', 770, None, None),
 (181508, 'Pastel Pinks on Dark Gray', 770, None, None),
 (181517, 'Rust, Pink on Yellow', 770, None, None),
 (181517, 'Rust, Pink on Yellow', 770, None, None),
 (181526, 'Muted Green, Dark Brown, Pink Stripe', 770, None, None),
 (181526, 'Muted Green, Dark Brown, Pink Stripe', 770, None, None),
 (181533, 'Pink, Yellow and White Center', 770, None, None),
 (181533, 'Pink, Yellow and White Center', 770, None, None),
 (181540, 'Untitled III', 770, None, None),
 (181540, 'Untitled III', 770, None, None),
 (181547, 'White Center, Gold and Gray on Maroon Red', 770, None, None),
 (181547, 'White Center, Gold and Gray on Maroon Red', 770, None, None),
 (181555, 'Gray, Light Brown on Brush Pink', 770, None, None),
 (181555, 'Gray, Light Brown on Brush Pink', 770, None, None),
 (181562, 'Orange on Orange', 770, None, None),
 (181562, 'Orange on Orange', 770, None, None),
 (181569, 'Yellow and Teal Blue', 770, None, None),
 (181569, 'Yellow and Teal Blue', 770, None, None),
 (181576, 'Black and White and Grey', 770, None, None),
 (181576, 'Black and White and Grey', 770, None, None),
 (181585, 'Black, White Stripe, Green On Pink', 770, None, None),
 (181585, 'Black, White Stripe, Green On Pink', 770, None, None),
 (181594, 'Pastel Shades on Orange', 770, None, None),
 (181594, 'Pastel Shades on Orange', 770, None, None),
 (181743, 'Gray Stripe, Yellow and White', 770, None, None),
 (181743, 'Gray Stripe, Yellow and White', 770, None, None),
 (181752, 'Lavender and Pastel Colors', 770, None, None)

All of the duplicates are **exact duplicates**. One record appears three times while all the other two. We'll follow the **same procedure** as for the subject table, after we find the odd artist_id and delete records with museum_id = 30.

In [45]:
%%sql

SELECT MIN(artist_id), MAX(artist_id) FROM work;

 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[(500, 920)]

It seems that the **artist_id = 500** is the value that is **not included in the artist table**, so let's drop that too.

In [46]:
%%sql

DELETE FROM WORK
WHERE artist_id = 500;

DELETE FROM work
WHERE museum_id = 30;

CREATE TABLE work_clean AS 
SELECT DISTINCT ON (work_id) * 
FROM work
ORDER BY work_id;

DROP TABLE work;

ALTER TABLE work_clean RENAME TO work;

 * postgresql://postgres:***@localhost:5432/famous_paintings
469 rows affected.
53 rows affected.
14193 rows affected.
Done.
Done.


[]

Time to start analysing!!!

# Analyse Data
## Featured artists
We'll first have a look at the artists with the most paintings in this list and then check if their paintings' regular prices aggregate to the greatest value or there are other artists with less featured paintings have a more valuable portfolio in this list.
### Total Number of Paintings featured

In [47]:
%%sql

SELECT a.full_name, COUNT(*) AS "Total Number of Paintings Featured"
FROM artist a
JOIN work w 
ON a.artist_id = w.artist_id
JOIN product_size p
ON w.work_id = p.work_id
GROUP BY a.artist_id, a.full_name
ORDER BY COUNT(*) DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/famous_paintings
10 rows affected.


[('Claude Monet', 2828),
 ('Vincent Van Gogh', 2535),
 ('Albert Marquet', 2034),
 ('Maurice Utrillo', 1876),
 ('Henri Lebasque', 1646),
 ('Louis Valtat', 1580),
 ('Camille Pissarro', 1562),
 ('Maximilien Luce', 1419),
 ('Henri Le Sidaner', 1375),
 ('Paul Cézanne', 1277)]

Claude Monet's and Vincent Van Gogh's paintings are featured much more than other artists' and as they are also very much celebrated, I guess they will top the list of those with the greatest sum of regular prices of their paintings, but let's confirm that assumption.  
### Sum of Regular Value of Paintings

In [48]:
%%sql

SELECT a.full_name, SUM(p.regular_price) AS "Sum of Regular Value of Paintings"
FROM artist a
JOIN work w 
ON a.artist_id = w.artist_id
JOIN product_size p
ON w.work_id = p.work_id
GROUP BY a.artist_id, a.full_name
ORDER BY SUM(p.regular_price) DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/famous_paintings
10 rows affected.


[('Claude Monet', 1909780),
 ('Vincent Van Gogh', 1527555),
 ('Albert Marquet', 1441830),
 ('Maurice Utrillo', 1263100),
 ('Henri Lebasque', 1191760),
 ('Louis Valtat', 1133380),
 ('Camille Pissarro', 1053210),
 ('Maximilien Luce', 1036645),
 ('Henri Le Sidaner', 982185),
 ('Pierre Bonnard', 831995)]

The only difference from the previous top 10 is that Pierre Bonnard entered the list and Paul Cezanne fell back, which raises another question... Who sells at a higher price on average?  
### Average Price per Painting

In [49]:
%%sql

SELECT a.full_name, ROUND(AVG(p.regular_price)) AS "Average Price per Painting"
FROM artist a
JOIN work w 
ON a.artist_id = w.artist_id
JOIN product_size p
ON w.work_id = p.work_id
GROUP BY a.artist_id, a.full_name
ORDER BY AVG(p.regular_price) DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/famous_paintings
10 rows affected.


[('Alfred Thompson Bricher', Decimal('1061')),
 ('Gustave Doré', Decimal('1033')),
 ('Vasiliy Polenov', Decimal('1009')),
 ('Buonarroti Michelangelo', Decimal('1008')),
 ('Edward Burne Jones', Decimal('1001')),
 ('Nicholas Pocock', Decimal('991')),
 ('Thomas Moran', Decimal('987')),
 ('Ferdinand Bol', Decimal('984')),
 ('Charles Marion Russell', Decimal('982')),
 ('Dominic Serres', Decimal('982'))]

None of the top selling artists appears in this list! Although it did come as a surprise to me, I should have expected it as rarity adds value to paintings and collectors pay substantial amount of money to acquire them. But are these the painters that have the highest valued products (regular price) in the dataset? I'm going to answer that question just after examining another rarity variable: age of death! Could early death of an artist add value to his/her paintings or someone who's lived a long life has become more of a master and his works are valued more?
### Age of Death Effect
We'll divide paintings into 3 categories according to the age of death of their creator and compare the average regular price of each category to look for any significant difference in values.

In [50]:
%%sql

SELECT CASE 
            WHEN a.death - a.birth < 30 THEN 'Young (<30)' 
            WHEN a.death - a.birth > 80 then 'Old (>80)'
            ELSE 'Too old to rock n'' roll, too young to die!' END AS "Age of Death of Painter" , 
            COUNT(*),
            ROUND(AVG(p.regular_price)) AS "Average Price per Product"
FROM artist a
JOIN work w 
ON a.artist_id = w.artist_id
JOIN product_size p
ON w.work_id = p.work_id
GROUP BY "Age of Death of Painter";

 * postgresql://postgres:***@localhost:5432/famous_paintings
3 rows affected.


[('Young (<30)', 1143, Decimal('759')),
 ("Too old to rock n' roll, too young to die!", 82896, Decimal('749')),
 ('Old (>80)', 21381, Decimal('746'))]

There's only a slight but existing trend for positive difference in the average price of paintings made by early departed artists. I picked <30 as an upper limit to determine 'Young' but any value up to <36 confirms that trend.  Let's also check the minimum age of painters in the list.
### Minimum Age Needed to Reach Be Famous Painter

In [51]:
%%sql

WITH cte AS (
    SELECT full_name, 
    death - birth AS "Age of Death", 
    RANK() OVER(ORDER BY(death - birth)) AS rnk
FROM artist
)
SELECT full_name, "Age of Death"
FROM cte
WHERE rnk = 1;

 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[('August Macke', 27)]

No painter made it into the list before he got 27.  
Now back to examine if those who consistently make highly valued works, are the ones with the highest priced paintings of all.
### Highest Valued Paintings

In [52]:
%%sql

SELECT a.full_name, w.work_id, w.name, s.subject, p.size_id, p.regular_price AS "Highest Valued Paintings Price"
FROM artist a
JOIN work w 
ON a.artist_id = w.artist_id
JOIN product_size p
ON w.work_id = p.work_id
LEFT JOIN subject s
ON w.work_id = s.work_id
ORDER BY p.regular_price DESC
LIMIT 20;

 * postgresql://postgres:***@localhost:5432/famous_paintings
20 rows affected.


[('Maximilien Luce', 133971, 'A Street in Paris in May 1871', None, '#VALUE!', 2235),
 ('Claude Monet', 2482, 'The Luncheon', 'Gardens', '79.163', 2055),
 ('William Adolphe Bouguereau', 4991, 'The Lost Pleiad', 'Nude', '4896', 2045),
 ('William Adolphe Bouguereau', 5012, 'The Storm', 'Seascapes', '4896', 2045),
 ('William Adolphe Bouguereau', 4805, 'Equality Before Death', 'Nude', '9648', 2045),
 ('Sir Lawrence Alma Tadema', 20958, 'God Speed!', None, '4896', 2045),
 ('Peter Paul Rubens', 22890, 'Fortuna', 'Nude', '4896', 2045),
 ('William Adolphe Bouguereau', 4773, 'Before the Bath (Avant le bain)', None, '4896', 2045),
 ('William Adolphe Bouguereau', 4791, 'Compassion', None, '4896', 2045),
 ('William Adolphe Bouguereau', 4757, 'Art and Literature', None, '4896', 2045),
 ('William Adolphe Bouguereau', 4799, 'Dawn (Reduction)', 'Nude', '4896', 2045),
 ('William Adolphe Bouguereau', 4753, 'After the Bath (Apres le Bain)', 'Nude', '4896', 2045),
 ('Sir Lawrence Alma Tadema', 20936, 'A Reading from Homer', None, '9648', 2045),
 ('William Adolphe Bouguereau', 4807, 'Evening Mood (La Crepuscule)', 'Nude', '4896', 2045),
 ('William Adolphe Bouguereau', 4969, 'The Ford (Le Gue)', None, '4896', 2045),
 ('Maximilien Luce', 133520, 'Les Batteurs de Pieux Entre', None, '#VALUE!', 2015),
 ('Franz Xavier Winterhalter', 29153, 'The Maharajah Duleep Singh', 'Winter', '4896', 1985),
 ('Leonardo Da Vinci', 1960, 'The Last Supper', 'Jesus Christ', '9648', 1985),
 ('Jean Honoré Fragonard', 26187, 'Winter', 'Winter', '9648', 1985),
 ('John Frederick Herring', 152491, 'The Doncaster Gold Cup 1825', 'Horses', '9648', 1985)]

Even if Maximilien Luce's 'A Street in Paris in May 1871' holds the first place in this list, the most striking fact is that Bouguereau has 10 paintings in the top 20. We can also notice that 17 out of the top 20 paintings are 48"x96" or 96"x48". Another thing here is that (except maybe the first place of the list) none of these paintings seems to be valued at an extremely high price. That means that either there are no outliers in the regular price column, or all of these prices are outliers! Let's find out...
### Regular Price Outliers

In [53]:
%%sql

WITH price_stats AS (
    SELECT 
        AVG(regular_price) AS mean_price,
        STDDEV(regular_price) AS stddev_price
    FROM product_size
)
SELECT p.*
FROM product_size p
JOIN price_stats ps ON TRUE
WHERE ABS((p.regular_price - ps.mean_price) / ps.stddev_price) > 3
ORDER BY p.regular_price DESC;

 * postgresql://postgres:***@localhost:5432/famous_paintings
195 rows affected.


[(133971, '#VALUE!', 1025, 2235),
 (2482, '79.163', 1075, 2055),
 (5012, '4896', 1055, 2045),
 (20958, '4896', 1055, 2045),
 (20936, '9648', 1055, 2045),
 (22890, '4896', 1115, 2045),
 (4969, '4896', 1055, 2045),
 (4991, '4896', 1055, 2045),
 (4757, '4896', 1055, 2045),
 (4773, '4896', 1055, 2045),
 (4791, '4896', 1055, 2045),
 (4799, '4896', 1055, 2045),
 (4805, '9648', 1085, 2045),
 (4807, '4896', 1055, 2045),
 (4753, '4896', 1055, 2045),
 (133520, '#VALUE!', 965, 2015),
 (29153, '4896', 1025, 1985),
 (167872, '9648', 1055, 1985),
 (152491, '9648', 1055, 1985),
 (26187, '9648', 1025, 1985),
 (1960, '9648', 1055, 1985),
 (194030, '9648', 1025, 1975),
 (20560, '9648', 1055, 1975),
 (149162, '4896', 1025, 1975),
 (19087, '4896', 1055, 1975),
 (27587, '4896', 1025, 1975),
 (27129, '9648', 1025, 1975),
 (24357, '4896', 1025, 1975),
 (24033, '9648', 1025, 1975),
 (23935, '9648', 1025, 1975),
 (5694, '4896', 1025, 1975),
 (22754, '4896', 1025, 1975),
 (22751, '4896', 1025, 1975),
 (21048, '9648', 1025, 1975),
 (21044, '9648', 1025, 1975),
 (21038, '9648', 1025, 1975),
 (7881, '9648', 1085, 1975),
 (169345, '9648', 1075, 1975),
 (207310, '9648', 1005, 1945),
 (7559, '4896', 1005, 1945),
 (7837, '9648', 1005, 1945),
 (7925, '9648', 1005, 1945),
 (7943, '9648', 1005, 1945),
 (7945, '9648', 1005, 1945),
 (17712, '4896', 1005, 1945),
 (18553, '9648', 1005, 1945),
 (18557, '9648', 1005, 1945),
 (18689, '9648', 1005, 1945),
 (20048, '9648', 1005, 1945),
 (20060, '9648', 1005, 1945),
 (20070, '9648', 1005, 1945),
 (20170, '9648', 1005, 1945),
 (23027, '9648', 1005, 1945),
 (23039, '9648', 1005, 1945),
 (23051, '9648', 1005, 1945),
 (23225, '9648', 1005, 1945),
 (23357, '9648', 1005, 1945),
 (23359, '9648', 1005, 1945),
 (23402, '9648', 1005, 1945),
 (27958, '9648', 1005, 1945),
 (27962, '9648', 1005, 1945),
 (28491, '4896', 1005, 1945),
 (28519, '9648', 1005, 1945),
 (28543, '4896', 1005, 1945),
 (28607, '4896', 1005, 1945),
 (28635, '4896', 1005, 1945),
 (28955, '9648', 1005, 1945),
 (31714, '4896', 1005, 1945),
 (150123, '9648', 1005, 1945),
 (150132, '9648', 1005, 1945),
 (150141, '9648', 1005, 1945),
 (170583, '9648', 1005, 1945),
 (185403, '9648', 1005, 1945),
 (185411, '9648', 1005, 1945),
 (185419, '9648', 1005, 1945),
 (185427, '9648', 1005, 1945),
 (185435, '9648', 1005, 1945),
 (185443, '9648', 1005, 1945),
 (185451, '9648', 1005, 1945),
 (185459, '9648', 1005, 1945),
 (185467, '9648', 1005, 1945),
 (185475, '9648', 1005, 1945),
 (185483, '9648', 1005, 1945),
 (185491, '9648', 1005, 1945),
 (185499, '9648', 1005, 1945),
 (185507, '9648', 1005, 1945),
 (185515, '9648', 1005, 1945),
 (185523, '9648', 1005, 1945),
 (185531, '9648', 1005, 1945),
 (185539, '9648', 1005, 1945),
 (185547, '9648', 1005, 1945),
 (185555, '9648', 1005, 1945),
 (185563, '9648', 1005, 1945),
 (185571, '9648', 1005, 1945),
 (185579, '9648', 1005, 1945),
 (185587, '9648', 1005, 1945),
 (185595, '9648', 1005, 1945),
 (185603, '9648', 1005, 1945),
 (185611, '9648', 1005, 1945),
 (185619, '9648', 1005, 1945),
 (185627, '9648', 1005, 1945),
 (185635, '9648', 1005, 1945),
 (185643, '9648', 1005, 1945),
 (185651, '9648', 1005, 1945),
 (185736, '9648', 1005, 1945),
 (186116, '9648', 1005, 1945),
 (186621, '9648', 1005, 1945),
 (186630, '9648', 1005, 1945),
 (186639, '9648', 1005, 1945),
 (186648, '9648', 1005, 1945),
 (186657, '9648', 1005, 1945),
 (186928, '9648', 1005, 1945),
 (186936, '9648', 1005, 1945),
 (186944, '9648', 1005, 1945),
 (186952, '9648', 1005, 1945),
 (186960, '9648', 1005, 1945),
 (5996, '9648', 1025, 1935),
 (5988, '9648', 1005, 1935),
 (5786, '4896', 1005, 1935),
 (187033, '9648', 1005, 1935),
 (187042, '9648', 1005, 1935),
 (187051, '9648', 1005, 1935),
 (20008, '9648', 1005, 1935),
 (23299, '4896', 1005, 1935),
 (23307, '4896', 1005, 1935),
 (19515, '9648', 1005, 1935),
 (30290, '9648', 1005, 1935),
 (187060, '9648', 1005, 1935),
 (187015, '9648', 1005, 1935),
 (30274, '9648', 1005, 1935),
 (8079, '9648', 1005, 1935),
 (26071, '9648', 1005, 19

It seems that 195 records could be classified as outliers.  
Let's see who are the top 10 artists with the highest valued paintings.
### Top 10 Artists with the Highest Valued Paintings

In [54]:
%%sql

SELECT * FROM (
    SELECT DISTINCT ON (a.artist_id) 
        a.full_name, 
        w.work_id, 
        w.name, 
        s.subject, 
        p.size_id, 
        p.regular_price AS "Highest Valued Painting"
    FROM artist a
    JOIN work w ON a.artist_id = w.artist_id
    JOIN product_size p ON w.work_id = p.work_id
    LEFT JOIN subject s ON w.work_id = s.work_id
    ORDER BY a.artist_id, p.regular_price DESC
) sub
ORDER BY "Highest Valued Painting" DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/famous_paintings
10 rows affected.


[('Maximilien Luce', 133971, 'A Street in Paris in May 1871', None, '#VALUE!', 2235),
 ('Claude Monet', 2482, 'The Luncheon', 'Gardens', '79.163', 2055),
 ('Peter Paul Rubens', 22890, 'Fortuna', 'Nude', '4896', 2045),
 ('William Adolphe Bouguereau', 5012, 'The Storm', 'Seascapes', '4896', 2045),
 ('Sir Lawrence Alma Tadema', 20936, 'A Reading from Homer', None, '9648', 2045),
 ('Gerard Van Honthorst', 167872, 'Shepherd Playing the Flute, and Four Shepherdesses', None, '9648', 1985),
 ('Jean Honoré Fragonard', 26187, 'Winter', 'Winter', '9648', 1985),
 ('John Frederick Herring', 152491, 'The Doncaster Gold Cup 1825', 'Horses', '9648', 1985),
 ('Franz Xavier Winterhalter', 29153, 'The Maharajah Duleep Singh', 'Winter', '4896', 1985),
 ('Leonardo Da Vinci', 1960, 'The Last Supper', 'Jesus Christ', '9648', 1985)]

Next on, we'll search for the subjects featured in the highest average price paintings.
## Featured Subjects
### Highest Valued Subjects

In [55]:
%%sql

SELECT s.subject AS "Highest Valued Subjects", 
       ROUND(AVG(p.regular_price)) AS "Regular Price"
    FROM work w 
    JOIN product_size p 
    ON w.work_id = p.work_id
    LEFT JOIN subject s ON w.work_id = s.work_id
    GROUP BY s.subject
    ORDER BY AVG(p.regular_price) DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/famous_paintings
10 rows affected.


[('Landscape Art', Decimal('877')),
 ('Marine Art/Maritime', Decimal('839')),
 ('Architectures', Decimal('812')),
 ('Deers', Decimal('794')),
 ('Lovers', Decimal('791')),
 ('Abstract/Modern Art', Decimal('781')),
 ('Horses', Decimal('773')),
 ('Nude', Decimal('767')),
 ('Jesus Christ', Decimal('766')),
 ('Churches/Temples/Mosques', Decimal('764'))]

We can see that Horses and Jesus Christ are the 2 common subjects with the last list and the average price of those categories is less than half the price of the highest valued product of the corresponding subject.  
But which subjects are featured most?  
### Most Featured Subjects

In [56]:
%%sql

SELECT s.subject AS "Most Featured Subjects", 
       COUNT(*) AS "Count"
    FROM product_size p 
    LEFT JOIN subject s ON p.work_id = s.work_id
    GROUP BY s.subject
    ORDER BY "Count" DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/famous_paintings
10 rows affected.


[(None, 65471),
 ('Portraits', 7889),
 ('Nude', 3809),
 ('Flowers', 3360),
 ('Rivers/Lakes', 2979),
 ('Still-Life', 2698),
 ('Abstract/Modern Art', 2662),
 ('Landscape Art', 2360),
 ('Seascapes', 2094),
 ('Horses', 1935)]

While there are many NULL values in the subject column, the trend is obvious: Portraits are featured more than twice the times of the second category and Abstract/Modern Art and Landscape Art are both pop and high valued subjects.  
But let's also look at the Lowest Valued Subjects!  
### Lowest Valued Subjects

In [57]:
%%sql

SELECT s.subject AS "Lowest Valued Subjects", 
       ROUND(AVG(p.regular_price)) AS "Regular Price"
    FROM work w 
    JOIN product_size p 
    ON w.work_id = p.work_id
    JOIN subject s ON w.work_id = s.work_id
    GROUP BY s.subject
    ORDER BY AVG(p.regular_price)
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/famous_paintings
10 rows affected.


[('Still-Life', Decimal('671')),
 ('Flowers', Decimal('695')),
 ('Autumn/Fall', Decimal('716')),
 ('Bridges', Decimal('717')),
 ('Water Lilies', Decimal('728')),
 ('Cafes/Bars', Decimal('730')),
 ('Spring', Decimal('731')),
 ('Winter', Decimal('732')),
 ('Summer', Decimal('735')),
 ('Dogs', Decimal('736'))]

These categories are the Lowest Valued Subjects, so it's no wonder that none of them made it into the Most Featured Subjects list! Still Life and Flowers don't seem to be much appreciated, but this is probably a subjective matter. A more objective one is that bigger paintings require more material and time to complete, so they should be valued more than smaller ones, shouldn't they? 
## Featured Canvas Sizes
### Highest Average Painting Prices by Canvas Size

In [58]:
%%sql

SELECT  p.size_id, c.width, c.height, ROUND(AVG(p.regular_price)) AS "Average Painting Prices by Canvas Size"
    FROM product_size p
    JOIN canvas_size c
    ON p.size_id = c.size_id::TEXT
    GROUP BY p.size_id, c.width, c.height
    ORDER BY ROUND(AVG(p.regular_price)) DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/famous_paintings
10 rows affected.


[('4896', 48, 96, Decimal('1949')),
 ('9648', 96, 48, Decimal('1925')),
 ('5468', 54, 68, Decimal('1495')),
 ('6854', 68, 54, Decimal('1484')),
 ('6060', 60, 60, Decimal('1471')),
 ('7171', 71, 71, Decimal('1425')),
 ('4872', 48, 72, Decimal('1418')),
 ('7248', 72, 48, Decimal('1415')),
 ('4080', 40, 80, Decimal('1348')),
 ('8040', 80, 40, Decimal('1330'))]

The most notable things here are that the 48"x96" and 96"x48" are by far the highest valued sizes, landscape oriented paintings are valued a little more than portrait (at least at the big size range that we can see here) and while bigger sizes seem to be valued higher, 71"x71" size paintings are valued less (on average) than some smaller ones. But which canvas sizes are met most frequently in this paintings list?  
### Most Frequently Used Canvas Sizes

In [59]:
%%sql

SELECT  p.size_id, c.width, c.height, COUNT(*) AS "Number of Paintings"
    FROM product_size p
    LEFT JOIN canvas_size c
    ON p.size_id = c.size_id::TEXT
    GROUP BY p.size_id, c.width, c.height
    ORDER BY COUNT(*) DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/famous_paintings
10 rows affected.


[('3040', 30, 40, 4563),
 ('3648', 36, 48, 4563),
 ('2936', 29, 36, 4535),
 ('4860', 48, 60, 4466),
 ('4030', 40, 30, 4254),
 ('4836', 48, 36, 4252),
 ('3629', 36, 29, 4186),
 ('2632', 26, 32, 4167),
 ('6048', 60, 48, 4112),
 ('2430', 24, 30, 4053)]

The most frequently featured ones seem to be in the medium (or a bit smaller) size range.

## Featured Museums
While validating our data we got a first impression that most of the featured museums are in USA and France but let's investigate further.
### Number of Featured Museums by Country

In [60]:
%%sql 

SELECT country, COUNT(*) AS "Number of Featured Museums"
FROM museum
GROUP BY country
ORDER BY COUNT(*) DESC;

 * postgresql://postgres:***@localhost:5432/famous_paintings
17 rows affected.


[('USA', 24),
 ('France', 7),
 ('England', 5),
 ('Netherlands', 4),
 ('Russia', 2),
 ('Switzerland', 2),
 ('Spain', 2),
 ('Germany', 1),
 ('Japan', 1),
 ('Wales', 1),
 ('Czechia', 1),
 ('Italy', 1),
 ('Hungary', 1),
 ('Norway', 1),
 ('Israel', 1),
 ('Brazil', 1),
 ('Australia', 1)]

24 out of 56 museums are located in the USA. That certainly explains the bias towards the American artists of the dataset! There are lots  of French or Dutch or Italian world-wide known artists, but American? Not so many in my (I must admit mediocre and European - could be biased too!) Knowledge.  
An interesting question that comes to mind is at which percentage paintings are exhibited domestically, compared to internationally. For the museums of each country, I'll calculate the percentage of paintings of domestic to all painters and then compare that value with the corresponding value from all other countries' museums.
### National to International Exposure
Be warned, this is going to be a big, quite complicated query so I'll add a few comments.

In [61]:
%%sql

WITH museum_paintings AS (
    SELECT 
        m.country AS museum_country, 
        -- Turn artists' nationality into country of origin to directly compare with museum.country
        CASE
            WHEN a.nationality = 'American' THEN 'USA'
            WHEN a.nationality = 'English' THEN 'England'
            WHEN a.nationality = 'Russian' THEN 'Russia'
            WHEN a.nationality = 'Spanish' THEN 'Spain'
            WHEN a.nationality = 'French' THEN 'France'
            WHEN a.nationality = 'Dutch' THEN 'Netherlands'
            WHEN a.nationality = 'German' THEN 'Germany'
            ELSE 'Other'
        END AS artist_country, 
        w.work_id
    FROM work w
    JOIN artist a ON w.artist_id = a.artist_id
    JOIN museum m ON w.museum_id = m.museum_id
),

-- Total paintings in all museums
total_paintings_all AS (
    SELECT COUNT(*) AS total_paintings_global FROM museum_paintings
),

-- Domestic paintings (by local artists)
domestic_exhibition_counts AS (
    SELECT 
        museum_country,
        COUNT(*) AS total_paintings,
        COUNT(*) FILTER (WHERE museum_country = artist_country) AS domestic_paintings
    FROM museum_paintings
    GROUP BY museum_country
),

-- Foreign exhibitions (paintings by artists of a country that are exhibited abroad)
foreign_exhibition_counts AS (
    SELECT 
        artist_country AS country,
        COUNT(*) AS foreign_exhibitions
    FROM museum_paintings
    WHERE museum_country != artist_country
    GROUP BY artist_country
),

-- Total foreign paintings = Total paintings in all museums - Total paintings in museums of each country
total_foreign_exhibition_counts AS (
    SELECT 
        d.museum_country AS country,
        (tpa.total_paintings_global - d.total_paintings) AS total_foreign_paintings
    FROM domestic_exhibition_counts d
    CROSS JOIN total_paintings_all tpa
)

-- Final aggregation
SELECT 
    d.museum_country, 
    d.domestic_paintings, 
    d.total_paintings, 
    ROUND(100.0 * d.domestic_paintings / d.total_paintings, 1) AS domestic_pct,

    -- Use COALESCE to handle missing values for foreign exhibitions
    COALESCE(f.foreign_exhibitions, 0) AS foreign_exhibitions,
    
    -- Use COALESCE to handle missing values for total foreign paintings
    COALESCE(t.total_foreign_paintings, 0) AS total_foreign_paintings,
    
    -- Calculate international percentage, ensuring no division by zero, and replacing NULL with 0
    COALESCE(
        CASE 
            WHEN COALESCE(t.total_foreign_paintings, 0) > 0 
            THEN ROUND(100.0 * f.foreign_exhibitions / NULLIF(t.total_foreign_paintings, 0), 1)
            ELSE 0
        END,
        0
    ) AS international_pct
FROM domestic_exhibition_counts d
LEFT JOIN foreign_exhibition_counts f ON d.museum_country = f.country
LEFT JOIN total_foreign_exhibition_counts t ON d.museum_country = t.country
ORDER BY international_pct DESC, domestic_pct DESC;


 * postgresql://postgres:***@localhost:5432/famous_paintings
17 rows affected.


[('France', 272, 332, Decimal('81.9'), 1343, 3895, Decimal('34.5')),
 ('Netherlands', 392, 470, Decimal('83.4'), 379, 3757, Decimal('10.1')),
 ('England', 178, 630, Decimal('28.3'), 115, 3597, Decimal('3.2')),
 ('Spain', 83, 195, Decimal('42.6'), 108, 4032, Decimal('2.7')),
 ('USA', 691, 2409, Decimal('28.7'), 48, 1818, Decimal('2.6')),
 ('Germany', 9, 23, Decimal('39.1'), 106, 4204, Decimal('2.5')),
 ('Russia', 32, 156, Decimal('20.5'), 24, 4071, Decimal('0.6')),
 ('Norway', 0, 1, Decimal('0.0'), 0, 4226, Decimal('0')),
 ('Brazil', 0, 1, Decimal('0.0'), 0, 4226, Decimal('0')),
 ('Wales', 0, 1, Decimal('0.0'), 0, 4226, Decimal('0')),
 ('Czechia', 0, 1, Decimal('0.0'), 0, 4226, Decimal('0')),
 ('Israel', 0, 1, Decimal('0.0'), 0, 4226, Decimal('0')),
 ('Australia', 0, 1, Decimal('0.0'), 0, 4226, Decimal('0')),
 ('Italy', 0, 1, Decimal('0.0'), 0, 4226, Decimal('0')),
 ('Japan', 0, 1, Decimal('0.0'), 0, 4226, Decimal('0')),
 ('Switzerland', 0, 3, Decimal('0.0'), 0, 4224, Decimal('0')),
 ('Hungary', 0, 1, Decimal('0.0'), 0, 4226, Decimal('0'))]

Looking at the results we can say that internationally, French artists are the most renown (1 out of 3 paintings in museums outside France is French!) which also explains the high percentage (almost 82%) of French works in French museums. Dutch museums on the other hand, exhibit Dutch painters' works at an even higher rate while internationally they hold a respectable (but 3 times lower than the French) share of paintings of about 10%, something which reveals that they focus on their artists a lot more than the rest of the world.  
Back to more casual queries, let's find out which are the most frequent opening and closing hours (mode of open and close).
## Museum open/close hours
### Mode of Opening/Closing Times

In [62]:
%%sql

WITH mode_open AS (
    SELECT open AS most_frequent_time, COUNT(*) AS frequency 
    FROM museum_hours
    GROUP BY open
    ORDER BY COUNT(*) DESC
    LIMIT 1
),
mode_close AS (
    SELECT close AS most_frequent_time, COUNT(*) AS frequency 
    FROM museum_hours
    GROUP BY close
    ORDER BY COUNT(*) DESC
    LIMIT 1
)
SELECT 'Opening Time' AS time, most_frequent_time, frequency
       FROM mode_open
            UNION 
SELECT 'Closing Time' AS time, most_frequent_time, frequency
       FROM mode_close
ORDER BY time DESC;


 * postgresql://postgres:***@localhost:5432/famous_paintings
2 rows affected.


[('Opening Time', datetime.time(10, 0), 213),
 ('Closing Time', datetime.time(17, 0), 125)]

Most of the museums open at 10:00 and most close at 17:00. But which is the day most of them are open?
### Mode of Open Day(s)

In [63]:
%%sql

SELECT day, COUNT(*)
FROM museum_hours
GROUP BY day
ORDER BY COUNT(*) DESC;

 * postgresql://postgres:***@localhost:5432/famous_paintings
7 rows affected.


[('Saturday', 57),
 ('Friday', 56),
 ('Thursday', 56),
 ('Sunday', 55),
 ('Wednesday', 50),
 ('Tuesday', 42),
 ('Monday', 28)]

We've already seen that there are 56 unique museum_id values so something's definitely wrong with 57 occurencies of Saturday!

In [64]:
%%sql

SELECT museum_id, day, COUNT(*)
FROM museum_hours
GROUP BY museum_id, day
HAVING COUNT(*) > 1
ORDER BY COUNT(*) DESC;

 * postgresql://postgres:***@localhost:5432/famous_paintings
1 rows affected.


[(80, 'Saturday', 2)]

In [65]:
%%sql

SELECT *
FROM museum_hours
WHERE museum_id = 80;

 * postgresql://postgres:***@localhost:5432/famous_paintings
7 rows affected.


[(80, 'Sunday', datetime.time(11, 0), datetime.time(17, 0)),
 (80, 'Tuesday', datetime.time(11, 0), datetime.time(20, 0)),
 (80, 'Wednesday', datetime.time(11, 0), datetime.time(20, 0)),
 (80, 'Friday', datetime.time(11, 0), datetime.time(21, 0)),
 (80, 'Saturday', datetime.time(11, 0), datetime.time(17, 0)),
 (80, 'Saturday', datetime.time(11, 0), datetime.time(17, 0)),
 (80, 'Thursday', datetime.time(11, 0), datetime.time(17, 0))]

Found an exact duplicate! Let's get rid of it!

In [66]:
%%sql

CREATE TABLE museum_hours_clean AS 
SELECT DISTINCT ON (museum_id, day) * 
FROM museum_hours;

DROP TABLE museum_hours;

ALTER TABLE museum_hours_clean RENAME TO museum_hours;

 * postgresql://postgres:***@localhost:5432/famous_paintings
343 rows affected.
Done.
Done.


[]

In [67]:
%%sql

SELECT day, COUNT(*)
FROM museum_hours
GROUP BY day
ORDER BY COUNT(*) DESC;

 * postgresql://postgres:***@localhost:5432/famous_paintings
7 rows affected.


[('Saturday', 56),
 ('Friday', 56),
 ('Thursday', 56),
 ('Sunday', 55),
 ('Wednesday', 50),
 ('Tuesday', 42),
 ('Monday', 28)]

It seems that all museums are open Thursday to Saturday, there's only one that is closed on Sunday but half of them are closed on Monday.  
We could also find out which museums are open the most and least hours per week and how many those are.
### Museum Open for Most/Least Hours per Week

In [68]:
%%sql

WITH open_hours AS (
    SELECT museum_id,
           ROUND(EXTRACT(EPOCH FROM (close - open)) / 3600, 1) AS day_hours
    FROM museum_hours
)
SELECT name, "Hours Open per Week"
FROM (
    -- Get museum open for the most hours
    SELECT m.name,
           SUM(o.day_hours) AS "Hours Open per Week"
    FROM museum m
    JOIN open_hours o ON m.museum_id = o.museum_id
    GROUP BY m.name
    ORDER BY "Hours Open per Week" DESC
    LIMIT 1
) AS max_hours
UNION ALL
SELECT name, "Hours Open per Week"
FROM (
    -- Get museum open for the least hours
    SELECT m.name,
           SUM(o.day_hours) AS "Hours Open per Week"
    FROM museum m
    JOIN open_hours o ON m.museum_id = o.museum_id
    GROUP BY m.name
    ORDER BY "Hours Open per Week" ASC
    LIMIT 1
) AS min_hours;


 * postgresql://postgres:***@localhost:5432/famous_paintings
2 rows affected.


[('The Prado Museum', Decimal('69.0')),
 ('Norton Simon Museum', Decimal('27.0'))]

## Explore Bargains
### Largest Discounts

In [69]:
%%sql

SELECT *, 
       regular_price - sale_price AS discount
FROM product_size
WHERE sale_price < regular_price
ORDER BY discount DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/famous_paintings
10 rows affected.


[(133971, '#VALUE!', 1025, 2235, 1210),
 (133520, '#VALUE!', 965, 2015, 1050),
 (5012, '4896', 1055, 2045, 990),
 (4807, '4896', 1055, 2045, 990),
 (4991, '4896', 1055, 2045, 990),
 (4799, '4896', 1055, 2045, 990),
 (4753, '4896', 1055, 2045, 990),
 (4757, '4896', 1055, 2045, 990),
 (4773, '4896', 1055, 2045, 990),
 (4791, '4896', 1055, 2045, 990)]

The largest discounts come from high-end regular prices but which paintings could be the best investing options? Assuming that their real value is their regular_price we'll look into those with the best expected Return On Investment rate.
### Best Expected ROI Rates

In [70]:
%%sql

SELECT *, 
       regular_price - sale_price AS discount,
       ROUND((regular_price - sale_price)::NUMERIC * 100 / sale_price, 2) AS "Expected ROI %"
FROM product_size
WHERE sale_price < regular_price
ORDER BY "Expected ROI %" DESC
LIMIT 20;

 * postgresql://postgres:***@localhost:5432/famous_paintings
20 rows affected.


[(17351, '36', 10, 125, 115, Decimal('1150.00')),
 (31780, '36', 10, 125, 115, Decimal('1150.00')),
 (31780, '30', 10, 95, 85, Decimal('850.00')),
 (17351, '30', 10, 95, 85, Decimal('850.00')),
 (17351, '24', 10, 85, 75, Decimal('750.00')),
 (23710, '30', 20, 95, 75, Decimal('375.00')),
 (23710, '24', 20, 85, 65, Decimal('325.00')),
 (198417, '36', 30, 125, 95, Decimal('316.67')),
 (6546, '36', 30, 125, 95, Decimal('316.67')),
 (6504, '36', 30, 125, 95, Decimal('316.67')),
 (6558, '30', 30, 95, 65, Decimal('216.67')),
 (198417, '30', 30, 95, 65, Decimal('216.67')),
 (6546, '30', 30, 95, 65, Decimal('216.67')),
 (28295, '30', 30, 95, 65, Decimal('216.67')),
 (6504, '30', 30, 95, 65, Decimal('216.67')),
 (28287, '36', 40, 125, 85, Decimal('212.50')),
 (6558, '24', 30, 85, 55, Decimal('183.33')),
 (31974, '24', 30, 85, 55, Decimal('183.33')),
 (6546, '24', 30, 85, 55, Decimal('183.33')),
 (28295, '24', 30, 85, 55, Decimal('183.33'))]

Obviously, the greatest expected ROI rates come from low-end regular priced paintings of our dataset.
# Conlusion
This was a project to practise PostgreSQL and showcase my cleaning, analysing and analytical thinking skills. I tried to answer questions that came naturally to my mind rather than inventing questions to answer with fancy queries. I'm sure you can find some interesting questions to answer too!  
I hope you enjoyed the process as much as I did!